{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "A class that wraps the the GO ontology that constructs a graph from the GO OBO file.\n",
    "It provides methods to construct the GO graph, compute various GO-related metrics,\n",
    "propagate terms, and plot GO subgraphs in various ways.\n",
    "\"\"\"\n",
    "\n",
    "from pathlib import Path\n",
    "from typing import Optional, Union, Tuple, Container, List, Callable\n",
    "\n",
    "import networkx as nx\n",
    "from networkx.drawing import layout\n",
    "from matplotlib import pyplot as plt\n",
    "from goatools.obo_parser import GODag\n",
    "\n",
    "\n",
    "MF_ROOT_ID = 'GO:0003674'\n",
    "CC_ROOT_ID = 'GO:0005575'\n",
    "BP_ROOT_ID = 'GO:0008150'\n",
    "\n",
    "\n",
    "class GOGraph:\n",
    "    \"\"\"\n",
    "    A wrapper for the the GO ontology that constructs a graph from the GO OBO file.\n",
    "    It provides methods to construct the GO graph, compute various GO-related metrics,\n",
    "    propagate terms, and plot GO subgraphs in various ways.\n",
    "\n",
    "    Args:\n",
    "        go_path: The path to the GO OBO file.\n",
    "        root_ids: The root IDs of the GO graph.\n",
    "        discard_go_dag: Whether to discard the GOATools GODag object after construction.\n",
    "            If True, the GODag object will not be available after construction.\n",
    "            If False, the GODag object will be available as self.go_dag.\n",
    "    \"\"\"\n",
    "    def __init__(self, go_fpath: Union[Path, str],\n",
    "                 root_ids: Optional[Tuple[str, str, str]] = (MF_ROOT_ID, CC_ROOT_ID, BP_ROOT_ID),\n",
    "                 discard_go_dag: bool = False):\n",
    "        self.go_fpath = go_fpath\n",
    "        go_dag = GODag(go_fpath)\n",
    "        self.ids_to_terms = {go_id: term_node.name for go_id, term_node in go_dag.items()}\n",
    "        self.root_ids = root_ids\n",
    "        self.go_graph = self.construct_graph_from_godag(go_dag)\n",
    "        if not discard_go_dag:\n",
    "            self.go_dag = go_dag\n",
    "\n",
    "    def construct_graph_from_godag(self,\n",
    "        godag: GODag,\n",
    "        max_depth: Optional[int] = None,\n",
    "        direction: str = 'both'\n",
    "    ) -> nx.DiGraph:\n",
    "        \"\"\"\n",
    "        Construct a NetworkX directed graph from a GOATools GODag object.\n",
    "        \n",
    "        Args:\n",
    "            godag: A GOATools GODag object containing the Gene Ontology hierarchy\n",
    "            root_node: Optional root node ID to start subgraph construction from\n",
    "            max_depth: Optional maximum depth to traverse from root node\n",
    "            direction: Direction to traverse from root ('up', 'down', or 'both')\n",
    "            \n",
    "        Returns:\n",
    "            A NetworkX directed graph with both 'PARENT' and 'CHILD' edges\n",
    "        \"\"\"\n",
    "        G = nx.DiGraph()\n",
    "        queue = []\n",
    "        visited = set()\n",
    "\n",
    "        def add_node_and_edges(go_id: str, current_depth: int = 0):\n",
    "            \"\"\"Helper function to add nodes and edges\"\"\"\n",
    "            if go_id not in godag or go_id in visited:\n",
    "                return\n",
    "\n",
    "            term = godag[go_id]\n",
    "            # print(term.id, term.name)\n",
    "            visited.add(go_id)\n",
    "\n",
    "            # Add the node if not already present\n",
    "            # print('Checking if node exists', go_id, go_id in G)\n",
    "            if go_id not in G:\n",
    "                # print('Adding node', go_id)\n",
    "                G.add_node(go_id,\n",
    "                        name=term.name,\n",
    "                        namespace=term.namespace,\n",
    "                        level=term.level,\n",
    "                        depth=term.depth)\n",
    "\n",
    "            # If we've reached max depth, stop traversing\n",
    "            if max_depth is not None and current_depth >= max_depth:\n",
    "                return\n",
    "\n",
    "            # Add parent relationships (traverse up)\n",
    "            if direction in ['up', 'both']:\n",
    "                for parent in term.parents:\n",
    "                    if parent.id not in visited:\n",
    "                        queue.append((parent.id, current_depth + 1))\n",
    "                        G.add_node(go_id,\n",
    "                                name=parent.name,\n",
    "                                namespace=parent.namespace,\n",
    "                                level=parent.level,\n",
    "                                depth=parent.depth)\n",
    "                    # print('Adding edge', go_id, parent.id)\n",
    "                    G.add_edge(go_id, parent.id, type='PARENT')\n",
    "                    G.add_edge(parent.id, go_id, type='CHILD')\n",
    "\n",
    "            # Add child relationships (traverse down)\n",
    "            if direction in ['down', 'both']:\n",
    "                for child in term.children:\n",
    "                    if child.id not in visited:\n",
    "                        queue.append((child.id, current_depth + 1))\n",
    "                        G.add_node(child.id,\n",
    "                                name=child.name,\n",
    "                                namespace=child.namespace,\n",
    "                                level=child.level,\n",
    "                                depth=child.depth)\n",
    "                    G.add_edge(go_id, child.id, type='CHILD')\n",
    "                    G.add_edge(child.id, go_id, type='PARENT')\n",
    "\n",
    "        for go_id in self.root_ids:\n",
    "            queue.append((go_id, 0))\n",
    "\n",
    "        while queue:\n",
    "            go_id, current_depth = queue.pop(0)\n",
    "            add_node_and_edges(go_id, current_depth)\n",
    "\n",
    "        return G\n",
    "\n",
    "    def compute_excess_components(self,\n",
    "                                  predicted_terms: Container[str],\n",
    "                                  expected_num_components: Union[int, None]=None) -> int:\n",
    "        \"\"\"\n",
    "        Computes the number of excess disconnected components in the predicted subgraph.\n",
    "\n",
    "        Args:\n",
    "            predicted_terms: A container of GO IDs representing the predicted terms.\n",
    "            expected_num_components: The expected number of components. If None the expected number of components \n",
    "                is the number of root IDs used to construct the graph.\n",
    "        \n",
    "        Returns:\n",
    "            The number of excess disconnected components in the predicted subgraph.\n",
    "        \"\"\"\n",
    "        if expected_num_components is None:\n",
    "            expected_num_components = len(self.root_ids)\n",
    "        predicted_subgraph = self.go_graph.subgraph(predicted_terms)\n",
    "        count_weak = nx.number_weakly_connected_components(predicted_subgraph)\n",
    "        return max(0, count_weak - expected_num_components)\n",
    "\n",
    "    def compute_excess_components_per_term(self,\n",
    "                                           predicted_terms: Container[str],\n",
    "                                           expected_num_components: Union[int, None]=None) -> float:\n",
    "        \"\"\"\n",
    "        Computes the number of excess disconnected components in the predicted subgraph normalized by the number of terms.\n",
    "\n",
    "        Args:\n",
    "            predicted_terms: A container of GO IDs representing the predicted terms.\n",
    "            expected_num_components: The expected number of components. If None the expected number of components \n",
    "                is the number of root IDs used to construct the graph.\n",
    "\n",
    "        Returns:\n",
    "            The number of excess disconnected components in the predicted subgraph normalized by the number of terms.\n",
    "        \"\"\"\n",
    "        return self.compute_excess_components(predicted_terms=predicted_terms,\n",
    "                                            expected_num_components=expected_num_components) / len(predicted_terms)\n",
    "\n",
    "    @staticmethod\n",
    "    def compute_depths(G: nx.DiGraph, root_id: str) -> List[int]:\n",
    "        \"\"\"\n",
    "        Returns a list of depth values for each node in the graph.\n",
    "        The depth of a node is the number of edges from the root to the node as found by breadth-first search \n",
    "            (this may be different from the \"depth\" attribute of the nodes).\n",
    "        The root is the node with the given root_id.\n",
    "\n",
    "        Args:\n",
    "            G: A NetworkX directed graph.\n",
    "            root_id: The ID of the root node to start the breadth-first search from.\n",
    "\n",
    "        Returns:\n",
    "            A list of depth values for each node in the graph.\n",
    "        \"\"\"\n",
    "        depths = []\n",
    "        queue = [(root_id, 0)]\n",
    "        while queue:\n",
    "            node_id, depth = queue.pop(0)\n",
    "            depths.append(depth)\n",
    "            for _, dest, data in G.edges(node_id, data=True):\n",
    "                if data['type'] == 'CHILD':\n",
    "                    queue.append((dest, depth + 1))\n",
    "        return depths\n",
    "\n",
    "    def propagate_terms(self, source_terms: Container[str], direction: str='up') -> List[str]:\n",
    "        \"\"\"\n",
    "        Returns a list of node ids representing the subgraph induced by following edges from the source terms.\n",
    "        If the direction is 'up', only 'PARENT' edges are followed.\n",
    "        If the direction is 'down', only 'CHILD' edges are followed.\n",
    "        If the direction is 'both', both 'PARENT' and 'CHILD' edges are followed.\n",
    "\n",
    "        Args:\n",
    "            source_terms: A container of GO IDs representing the source terms.\n",
    "            direction: The direction to propagate the terms. Allowed values are \"up\" and \"down\" and \"both\"\n",
    "\n",
    "        Returns:\n",
    "            A list of GO IDs representing the propagated terms.\n",
    "        \"\"\"\n",
    "        if direction == 'up':\n",
    "            edge_types = ('PARENT',)\n",
    "        elif direction == 'down':\n",
    "            edge_types = ('CHILD',)\n",
    "        elif direction == 'both':\n",
    "            edge_types = ('PARENT', 'CHILD')\n",
    "        else:\n",
    "            raise ValueError(f'Invalid direction: {direction}. Allowed directions are \"up\", \"down\", and \"both\".')\n",
    "\n",
    "        propagated_nodes = []\n",
    "\n",
    "        for term in source_terms:\n",
    "            parent_queue = [term]\n",
    "\n",
    "            while parent_queue:\n",
    "                this_node = parent_queue.pop()\n",
    "                propagated_nodes.append(this_node)\n",
    "                for dest_node, edge_data in self.go_graph[this_node].items():\n",
    "                    if edge_data['type'] in edge_types:\n",
    "                        parent_queue.append(dest_node)\n",
    "        return propagated_nodes\n",
    "\n",
    "    @staticmethod\n",
    "    def plot_go_subgraph(G: nx.DiGraph,\n",
    "                           ax: Optional[plt.Axes]=None,\n",
    "                           label_by_name: bool=False,\n",
    "                           layout_func: Callable=layout.spring_layout,\n",
    "                           fig_width: float=12,\n",
    "                           node_size: float=50,\n",
    "                           edge_width: float=0.5,\n",
    "                           node_color: str='lightblue',\n",
    "                           edge_color: str='gray',\n",
    "                           label_size: float=6,\n",
    "                           ) -> Tuple[plt.Figure, plt.Axes]:\n",
    "        '''\n",
    "        Plots a subgraph of G containing only the nodes in node_list and only 'CHILD' edges.\n",
    "        '''\n",
    "        pos = layout_func(G)\n",
    "        if ax is None:\n",
    "            fig, ax = plt.subplots(1, figsize=(fig_width, fig_width), dpi=300)\n",
    "        else:\n",
    "            fig = ax.get_figure()\n",
    "\n",
    "        nx.draw_networkx_nodes(\n",
    "            G=G,\n",
    "            pos=pos,\n",
    "            ax=ax,\n",
    "            node_color=node_color,\n",
    "            node_size=node_size\n",
    "        )\n",
    "        # Filter edges to only those labeled as 'CHILD'\n",
    "        child_edges = [\n",
    "            (u, v) for u, v, d in G.edges(data=True)\n",
    "            if d.get('type') == 'CHILD'\n",
    "        ]\n",
    "        nx.draw_networkx_edges(\n",
    "            G=G,\n",
    "            pos=pos,\n",
    "            edgelist=child_edges,\n",
    "            ax=ax,\n",
    "            edge_color=edge_color,\n",
    "            arrows=True,\n",
    "            width=edge_width\n",
    "        )\n",
    "        if label_by_name:\n",
    "            node_labels = {node: G.nodes[node]['name'] for node in G.nodes()}\n",
    "        else:\n",
    "            node_labels = {node: node for node in G.nodes()}\n",
    "        nx.draw_networkx_labels(\n",
    "            G=G,\n",
    "            pos=pos,\n",
    "            ax=ax,\n",
    "            labels=node_labels,\n",
    "            verticalalignment='bottom',\n",
    "            horizontalalignment='center',\n",
    "            font_size=label_size\n",
    "        )\n",
    "        ax.set_axis_off()\n",
    "        return fig, ax\n",
    "\n",
    "    def plot_go_subgraphs_true_vs_predicted(self,\n",
    "                                            true_terms: Container[str],\n",
    "                                            predicted_terms: Container[str],\n",
    "                                            ax: Optional[plt.Axes]=None,\n",
    "                                            label_by_name: bool=False,\n",
    "                                            layout_func: Callable=layout.spring_layout,\n",
    "                                            fig_width: float=8,\n",
    "                                            node_size: float=50,\n",
    "                                            edge_width: float=0.5,\n",
    "                                            label_size: float=6,\n",
    "                                            tp_color: str='green',\n",
    "                                            fp_color: str='red',\n",
    "                                            fn_color: str='blue',\n",
    "                                            edge_color: str='gray',\n",
    "                                            ) -> Tuple[plt.Figure, plt.Axes]:\n",
    "        '''\n",
    "        Plots a subgraph of the GO graph containing the nodes representing the true and predicted terms.\n",
    "        The nodes are colored green, red, and blue (default colors) for true positives, false positives,\n",
    "            and false negatives, respectively.\n",
    "        The edges are colored gray (default). Only the 'CHILD' edges are shown.\n",
    "        The nodes are labeled by their GO IDs, or optionally by the term names.\n",
    "        \n",
    "        Args:\n",
    "            true_terms: A container of GO IDs representing the true terms.\n",
    "            predicted_terms: A container of GO IDs representing the predicted terms.\n",
    "            ax: An optional matplotlib Axes object to plot the graph on.\n",
    "            label_by_name: Whether to label the nodes by the term names instead of their GO IDs.\n",
    "            layout_func: A function to layout the graph.\n",
    "            fig_width: The width of the figure.\n",
    "            node_size: The size of the nodes.\n",
    "            edge_width: The width of the edges.\n",
    "            label_size: The size of the labels.\n",
    "            \n",
    "        '''\n",
    "        true_terms = set(true_terms)\n",
    "        predicted_terms = set(predicted_terms)\n",
    "        tp_terms = true_terms.intersection(predicted_terms)\n",
    "        fp_terms = predicted_terms.difference(tp_terms)\n",
    "        fn_terms = true_terms.difference(tp_terms)\n",
    "        joint_terms = true_terms.union(predicted_terms)\n",
    "\n",
    "        joint_network = self.go_graph.subgraph(joint_terms)\n",
    "\n",
    "        pos = layout_func(joint_network)\n",
    "\n",
    "        if ax is None:\n",
    "            fig, ax = plt.subplots(1, figsize=(fig_width, fig_width), dpi=300)\n",
    "        else:\n",
    "            fig = ax.get_figure()\n",
    "\n",
    "        nx.draw_networkx_nodes(\n",
    "            G=self.go_graph.subgraph(tp_terms),\n",
    "            pos=pos,\n",
    "            ax=ax,\n",
    "            node_color=tp_color,\n",
    "            node_size=node_size\n",
    "        )\n",
    "\n",
    "        nx.draw_networkx_nodes(\n",
    "            G=self.go_graph.subgraph(fp_terms),\n",
    "            pos=pos,\n",
    "            ax=ax,\n",
    "            node_color=fp_color,\n",
    "            node_size=node_size\n",
    "        )\n",
    "\n",
    "        nx.draw_networkx_nodes(\n",
    "            G=self.go_graph.subgraph(fn_terms),\n",
    "            pos=pos,\n",
    "            ax=ax,\n",
    "            node_color=fn_color,\n",
    "            node_size=node_size\n",
    "        )\n",
    "\n",
    "        # Filter edges to only those labeled as 'CHILD'\n",
    "        child_edges = [\n",
    "            (u, v) for u, v, d in joint_network.edges(data=True)\n",
    "            if d.get('type') == 'CHILD'\n",
    "        ]\n",
    "        nx.draw_networkx_edges(\n",
    "            G=joint_network,\n",
    "            pos=pos,\n",
    "            edgelist=child_edges,\n",
    "            ax=ax,\n",
    "            edge_color=edge_color,\n",
    "            arrows=True,\n",
    "            width=edge_width\n",
    "        )\n",
    "        if label_by_name:\n",
    "            node_labels = {node: joint_network.nodes[node]['name'] for node in joint_network.nodes()}\n",
    "        else:\n",
    "            node_labels = {node: node for node in joint_network.nodes()}\n",
    "        nx.draw_networkx_labels(\n",
    "            G=joint_network,\n",
    "            pos=pos,\n",
    "            ax=ax,\n",
    "            labels=node_labels,\n",
    "            verticalalignment='bottom',\n",
    "            horizontalalignment='center',\n",
    "            font_size=label_size\n",
    "        )\n",
    "        ax.set_axis_off()\n",
    "        return fig, ax\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../cafa-5-protein-function-prediction/Train/go-basic.obo: fmt(1.2) rel(2023-01-01) 46,739 Terms\n"
     ]
    }
   ],
   "source": [
    "cafa5_go_path = '../cafa-5-protein-function-prediction/Train/go-basic.obo'\n",
    "godag = GODag(cafa5_go_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__class__',\n",
       " '__class_getitem__',\n",
       " '__contains__',\n",
       " '__delattr__',\n",
       " '__delitem__',\n",
       " '__dict__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__firstlineno__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__getitem__',\n",
       " '__getstate__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__ior__',\n",
       " '__iter__',\n",
       " '__le__',\n",
       " '__len__',\n",
       " '__lt__',\n",
       " '__module__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__or__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__reversed__',\n",
       " '__ror__',\n",
       " '__setattr__',\n",
       " '__setitem__',\n",
       " '__sizeof__',\n",
       " '__static_attributes__',\n",
       " '__str__',\n",
       " '__subclasshook__',\n",
       " '__weakref__',\n",
       " '_populate_relationships',\n",
       " '_populate_terms',\n",
       " '_set_level_depth',\n",
       " '_str_desc',\n",
       " 'clear',\n",
       " 'copy',\n",
       " 'data_version',\n",
       " 'draw_lineage',\n",
       " 'fromkeys',\n",
       " 'get',\n",
       " 'id2int',\n",
       " 'items',\n",
       " 'keys',\n",
       " 'label_wrap',\n",
       " 'load_obo_file',\n",
       " 'make_graph_pydot',\n",
       " 'make_graph_pygraphviz',\n",
       " 'paths_to_top',\n",
       " 'pop',\n",
       " 'popitem',\n",
       " 'query_term',\n",
       " 'setdefault',\n",
       " 'typedefs',\n",
       " 'update',\n",
       " 'update_association',\n",
       " 'values',\n",
       " 'version',\n",
       " 'write_dag']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(godag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../cafa-5-protein-function-prediction/Train/go-basic.obo', 'r') as f:\n",
    "    obo_contents = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "format-version: 1.2\n",
      "data-version: releases/2023-01-01\n",
      "subsetdef: chebi_ph7_3 \"Rhea list of ChEBI terms representing the major species at pH 7.3.\"\n",
      "subsetdef: gocheck_do_not_annotate \"Term not to be used for direct annotation\"\n",
      "subsetdef: gocheck_do_not_manually_annotate \"Term not to be used for direct manual annotation\"\n",
      "subsetdef: goslim_agr \"AGR slim\"\n",
      "subsetdef: goslim_aspergillus \"Aspergillus GO slim\"\n",
      "subsetdef: goslim_candida \"Candida GO slim\"\n",
      "subsetdef: goslim_chembl \"ChEMBL protein targets summary\"\n",
      "subsetdef: goslim_drosophila \"Drosophila GO slim\"\n",
      "subsetdef: goslim_flybase_ribbon \"FlyBase Drosophila GO ribbon slim\"\n",
      "subsetdef: goslim_generic \"Generic GO slim\"\n",
      "subsetdef: goslim_metagenomics \"Metagenomics GO slim\"\n",
      "subsetdef: goslim_mouse \"Mouse GO slim\"\n",
      "subsetdef: goslim_pir \"PIR GO slim\"\n",
      "subsetdef: goslim_plant \"Plant GO slim\"\n",
      "subsetdef: goslim_pombe \"Fission yeast GO slim\"\n",
      "subsetdef: goslim_synapse \"synapse GO slim\"\n",
      "subsetdef: goslim_yeast \"Yeast GO slim\"\n",
      "subsetdef: prokaryote_subset \"GO subset for prokaryotes\"\n",
      "synonymtypedef: syngo_official_label \"label approved by the SynGO project\"\n",
      "synonymtypedef: systematic_synonym \"Systematic synonym\" EXACT\n",
      "default-namespace: gene_ontology\n",
      "ontology: go\n",
      "\n",
      "[Term]\n",
      "id: GO:0000001\n",
      "name: mitochondrion inheritance\n",
      "namespace: biological_process\n",
      "def: \"The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton.\" [GOC:mcc, PMID:10873824, PMID:11389764]\n",
      "synonym: \"mitochondrial inheritance\" EXACT []\n",
      "is_a: GO:0048308 ! organelle inheritance\n",
      "is_a: GO:0048311 ! mitochondrion distribution\n",
      "\n",
      "[Term]\n",
      "id: GO:0000002\n",
      "name: mitochondrial genome maintenance\n",
      "namespace: biological_process\n",
      "def: \"The maintenance of the structure and integrity of the mitochondrial genome; includes replication and segregation of the mitochondrial chromosome.\" [GOC:ai, GOC:vw]\n",
      "is_a: GO:0007005 ! mitochondrion organization\n",
      "\n",
      "[Term]\n",
      "id: GO:0000003\n",
      "name: reproduction\n",
      "namespace: biological_process\n",
      "alt_id: GO:0019952\n",
      "alt_id: GO:0050876\n",
      "def: \"The production of new individuals that contain some portion of genetic material inherited from one or more parent organisms.\" [GOC:go_curators, GOC:isa_complete, GOC:jl, ISBN:0198506732]\n",
      "subset: goslim_agr\n",
      "subset: goslim_chembl\n",
      "subset: goslim_flybase_ribbon\n",
      "subset: goslim_pir\n",
      "subset: goslim_plant\n",
      "synonym: \"reproductive physiological process\" EXACT []\n",
      "xref: Wikipedia:Reproduction\n",
      "is_a: GO:0008150 ! biological_process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000005\n",
      "name: obsolete ribosomal chaperone activity\n",
      "namespace: molecular_function\n",
      "def: \"OBSOLETE. Assists in the correct assembly of ribosomes or ribosomal subunits in vivo, but is not a component of the assembled ribosome when performing its normal biological function.\" [GOC:jl, PMID:12150913]\n",
      "comment: This term was made obsolete because it refers to a class of gene products and a biological process rather than a molecular function.\n",
      "synonym: \"ribosomal chaperone activity\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0042254\n",
      "consider: GO:0044183\n",
      "consider: GO:0051082\n",
      "\n",
      "[Term]\n",
      "id: GO:0000006\n",
      "name: high-affinity zinc transmembrane transporter activity\n",
      "namespace: molecular_function\n",
      "def: \"Enables the transfer of zinc ions (Zn2+) from one side of a membrane to the other, probably powered by proton motive force. In high-affinity transport the transporter is able to bind the solute even if it is only present at very low concentrations.\" [TC:2.A.5.1.1]\n",
      "synonym: \"high affinity zinc uptake transmembrane transporter activity\" EXACT []\n",
      "synonym: \"high-affinity zinc uptake transmembrane transporter activity\" RELATED []\n",
      "is_a: GO:0005385 ! zinc ion transmembrane transporter activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000007\n",
      "name: low-affinity zinc ion transmembrane transporter activity\n",
      "namespace: molecular_function\n",
      "def: \"Enables the transfer of a solute or solutes from one side of a membrane to the other according to the reaction: Zn2+ = Zn2+, probably powered by proton motive force. In low-affinity transport the transporter is able to bind the solute only if it is present at very high concentrations.\" [GOC:mtg_transport, ISBN:0815340729]\n",
      "is_a: GO:0005385 ! zinc ion transmembrane transporter activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000008\n",
      "name: obsolete thioredoxin\n",
      "namespace: molecular_function\n",
      "alt_id: GO:0000013\n",
      "def: \"OBSOLETE. A small disulfide-containing redox protein that serves as a general protein disulfide oxidoreductase. Interacts with a broad range of proteins by a redox mechanism, based on the reversible oxidation of 2 cysteine thiol groups to a disulfide, accompanied by the transfer of 2 electrons and 2 protons. The net result is the covalent interconversion of a disulfide and a dithiol.\" [GOC:kd]\n",
      "comment: This term was made obsolete because it represents gene products.\n",
      "synonym: \"thioredoxin\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0003756\n",
      "consider: GO:0015036\n",
      "\n",
      "[Term]\n",
      "id: GO:0000009\n",
      "name: alpha-1,6-mannosyltransferase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the transfer of a mannose residue to an oligosaccharide, forming an alpha-(1->6) linkage.\" [GOC:mcc, PMID:2644248]\n",
      "synonym: \"1,6-alpha-mannosyltransferase activity\" EXACT []\n",
      "xref: Reactome:R-HSA-449718 \"Addition of a third mannose to the N-glycan precursor by ALG2\"\n",
      "is_a: GO:0000030 ! mannosyltransferase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000010\n",
      "name: trans-hexaprenyltranstransferase activity\n",
      "namespace: molecular_function\n",
      "alt_id: GO:0036422\n",
      "def: \"Catalysis of the reaction: (2E,6E)-farnesyl diphosphate + 4 isopentenyl diphosphate = 4 diphosphate + all-trans-heptaprenyl diphosphate.\" [PMID:9708911, RHEA:27794]\n",
      "synonym: \"all-trans-heptaprenyl-diphosphate synthase activity\" RELATED [EC:2.5.1.30]\n",
      "synonym: \"HepPP synthase activity\" RELATED [EC:2.5.1.30]\n",
      "synonym: \"heptaprenyl diphosphate synthase activity\" RELATED []\n",
      "synonym: \"heptaprenyl pyrophosphate synthase activity\" RELATED [EC:2.5.1.30]\n",
      "synonym: \"heptaprenyl pyrophosphate synthetase activity\" RELATED [EC:2.5.1.30]\n",
      "xref: EC:2.5.1.30\n",
      "xref: MetaCyc:TRANS-HEXAPRENYLTRANSTRANSFERASE-RXN\n",
      "xref: RHEA:27794\n",
      "is_a: GO:0004659 ! prenyltransferase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000011\n",
      "name: vacuole inheritance\n",
      "namespace: biological_process\n",
      "def: \"The distribution of vacuoles into daughter cells after mitosis or meiosis, mediated by interactions between vacuoles and the cytoskeleton.\" [GOC:mcc, PMID:10873824, PMID:14616069]\n",
      "is_a: GO:0007033 ! vacuole organization\n",
      "is_a: GO:0048308 ! organelle inheritance\n",
      "\n",
      "[Term]\n",
      "id: GO:0000012\n",
      "name: single strand break repair\n",
      "namespace: biological_process\n",
      "def: \"The repair of single strand breaks in DNA. Repair of such breaks is mediated by the same enzyme systems as are used in base excision repair.\" [PMID:18626472]\n",
      "is_a: GO:0006281 ! DNA repair\n",
      "\n",
      "[Term]\n",
      "id: GO:0000014\n",
      "name: single-stranded DNA endodeoxyribonuclease activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the hydrolysis of ester linkages within a single-stranded deoxyribonucleic acid molecule by creating internal breaks.\" [GOC:mah]\n",
      "synonym: \"single-stranded DNA specific endodeoxyribonuclease activity\" RELATED []\n",
      "synonym: \"ssDNA-specific endodeoxyribonuclease activity\" RELATED [GOC:mah]\n",
      "is_a: GO:0004520 ! DNA endonuclease activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000015\n",
      "name: phosphopyruvate hydratase complex\n",
      "namespace: cellular_component\n",
      "def: \"A multimeric enzyme complex, usually a dimer or an octamer, that catalyzes the conversion of 2-phospho-D-glycerate to phosphoenolpyruvate and water.\" [GOC:jl, ISBN:0198506732]\n",
      "subset: goslim_metagenomics\n",
      "synonym: \"enolase complex\" EXACT []\n",
      "is_a: GO:1902494 ! catalytic complex\n",
      "relationship: part_of GO:0005829 ! cytosol\n",
      "\n",
      "[Term]\n",
      "id: GO:0000016\n",
      "name: lactase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the reaction: lactose + H2O = D-glucose + D-galactose.\" [PMID:12023280]\n",
      "synonym: \"lactose galactohydrolase activity\" EXACT []\n",
      "xref: EC:3.2.1.108\n",
      "xref: MetaCyc:LACTASE-RXN\n",
      "xref: Reactome:R-HSA-189062 \"lactose + H2O => D-glucose + D-galactose\"\n",
      "xref: Reactome:R-HSA-5658001 \"Defective LCT does not hydrolyze Lac\"\n",
      "xref: RHEA:10076\n",
      "is_a: GO:0004553 ! hydrolase activity, hydrolyzing O-glycosyl compounds\n",
      "\n",
      "[Term]\n",
      "id: GO:0000017\n",
      "name: alpha-glucoside transport\n",
      "namespace: biological_process\n",
      "def: \"The directed movement of alpha-glucosides into, out of or within a cell, or between cells, by means of some agent such as a transporter or pore. Alpha-glucosides are glycosides in which the sugar group is a glucose residue, and the anomeric carbon of the bond is in an alpha configuration.\" [GOC:jl, ISBN:0198506732, PMID:9919658]\n",
      "is_a: GO:0042946 ! glucoside transport\n",
      "\n",
      "[Term]\n",
      "id: GO:0000018\n",
      "name: regulation of DNA recombination\n",
      "namespace: biological_process\n",
      "def: \"Any process that modulates the frequency, rate or extent of DNA recombination, a DNA metabolic process in which a new genotype is formed by reassortment of genes resulting in gene combinations different from those that were present in the parents.\" [GOC:go_curators, ISBN:0198506732]\n",
      "is_a: GO:0051052 ! regulation of DNA metabolic process\n",
      "relationship: regulates GO:0006310 ! DNA recombination\n",
      "\n",
      "[Term]\n",
      "id: GO:0000019\n",
      "name: regulation of mitotic recombination\n",
      "namespace: biological_process\n",
      "def: \"Any process that modulates the frequency, rate or extent of DNA recombination during mitosis.\" [GOC:go_curators]\n",
      "synonym: \"regulation of recombination within rDNA repeats\" NARROW []\n",
      "is_a: GO:0000018 ! regulation of DNA recombination\n",
      "relationship: regulates GO:0006312 ! mitotic recombination\n",
      "\n",
      "[Term]\n",
      "id: GO:0000020\n",
      "name: obsolete negative regulation of recombination within rDNA repeats\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. Any process that stops, prevents, or reduces the frequency, rate or extent of genetic recombination within the DNA of the genes coding for ribosomal RNA.\" [GOC:go_curators, ISBN:0198506732]\n",
      "comment: This term was made obsolete because it describes a substrate-specific process.\n",
      "synonym: \"negative regulation of recombination within rDNA repeats\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0045950\n",
      "\n",
      "[Term]\n",
      "id: GO:0000022\n",
      "name: mitotic spindle elongation\n",
      "namespace: biological_process\n",
      "alt_id: GO:1905121\n",
      "def: \"The cell cycle process in which the distance is lengthened between poles of the mitotic spindle. Mitotic spindle elongation begins during mitotic prophase and ends during mitotic anaphase B.\" [GOC:mtg_cell_cycle, GOC:vw, PMID:19686686]\n",
      "synonym: \"microtubule sliding involved in mitotic spindle elongation\" RELATED []\n",
      "synonym: \"spindle elongation during mitosis\" EXACT []\n",
      "is_a: GO:0051231 ! spindle elongation\n",
      "is_a: GO:1903047 ! mitotic cell cycle process\n",
      "relationship: part_of GO:0000070 ! mitotic sister chromatid segregation\n",
      "relationship: part_of GO:0007052 ! mitotic spindle organization\n",
      "\n",
      "[Term]\n",
      "id: GO:0000023\n",
      "name: maltose metabolic process\n",
      "namespace: biological_process\n",
      "def: \"The chemical reactions and pathways involving the disaccharide maltose (4-O-alpha-D-glucopyranosyl-D-glucopyranose), an intermediate in the catabolism of glycogen and starch.\" [GOC:jl, ISBN:0198506732]\n",
      "synonym: \"malt sugar metabolic process\" EXACT []\n",
      "synonym: \"malt sugar metabolism\" EXACT []\n",
      "synonym: \"maltose metabolism\" EXACT []\n",
      "is_a: GO:0005984 ! disaccharide metabolic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000024\n",
      "name: maltose biosynthetic process\n",
      "namespace: biological_process\n",
      "def: \"The chemical reactions and pathways resulting in the formation of the disaccharide maltose (4-O-alpha-D-glucopyranosyl-D-glucopyranose).\" [GOC:jl, ISBN:0198506732]\n",
      "synonym: \"malt sugar biosynthesis\" EXACT []\n",
      "synonym: \"malt sugar biosynthetic process\" EXACT []\n",
      "synonym: \"maltose anabolism\" EXACT []\n",
      "synonym: \"maltose biosynthesis\" EXACT []\n",
      "synonym: \"maltose formation\" EXACT []\n",
      "synonym: \"maltose synthesis\" EXACT []\n",
      "is_a: GO:0000023 ! maltose metabolic process\n",
      "is_a: GO:0046351 ! disaccharide biosynthetic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000025\n",
      "name: maltose catabolic process\n",
      "namespace: biological_process\n",
      "def: \"The chemical reactions and pathways resulting in the breakdown of the disaccharide maltose (4-O-alpha-D-glucopyranosyl-D-glucopyranose).\" [GOC:jl, ISBN:0198506732]\n",
      "synonym: \"malt sugar catabolic process\" EXACT []\n",
      "synonym: \"malt sugar catabolism\" EXACT []\n",
      "synonym: \"maltose breakdown\" EXACT []\n",
      "synonym: \"maltose degradation\" EXACT []\n",
      "synonym: \"maltose hydrolysis\" NARROW []\n",
      "xref: MetaCyc:MALTOSECAT-PWY\n",
      "is_a: GO:0000023 ! maltose metabolic process\n",
      "is_a: GO:0046352 ! disaccharide catabolic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000026\n",
      "name: alpha-1,2-mannosyltransferase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the transfer of a mannose residue to an oligosaccharide, forming an alpha-(1->2) linkage.\" [GOC:mcc, PMID:10521541]\n",
      "xref: Reactome:R-HSA-446187 \"ALG11 transfers the fourth and fifth Man to the N-glycan precursor\"\n",
      "xref: Reactome:R-HSA-446215 \"ALG9 transfers Man to N-glycan precursor (GlcNAc)2 (Man)6 (PP-Dol)1\"\n",
      "xref: Reactome:R-HSA-446216 \"ALG9 transfers Man to N-glycan precursor (GlcNAc)2 (Man)8 (PP-Dol)1\"\n",
      "xref: Reactome:R-HSA-4551297 \"Defective ALG11 does not transfer Man to the N-glycan precursor\"\n",
      "xref: Reactome:R-HSA-4720478 \"Defective ALG9 does not add the seventh mannose to the N-glycan precursor\"\n",
      "xref: Reactome:R-HSA-9035514 \"Defective ALG9 does not add the last mannose to the N-glycan precursor\"\n",
      "is_a: GO:0000030 ! mannosyltransferase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000027\n",
      "name: ribosomal large subunit assembly\n",
      "namespace: biological_process\n",
      "def: \"The aggregation, arrangement and bonding together of constituent RNAs and proteins to form the large ribosomal subunit.\" [GOC:jl, PMID:30467428]\n",
      "synonym: \"50S ribosomal subunit assembly\" NARROW [GOC:mah]\n",
      "synonym: \"60S ribosomal subunit assembly\" NARROW [GOC:mah]\n",
      "is_a: GO:0022618 ! ribonucleoprotein complex assembly\n",
      "relationship: part_of GO:0042255 ! ribosome assembly\n",
      "relationship: part_of GO:0042273 ! ribosomal large subunit biogenesis\n",
      "\n",
      "[Term]\n",
      "id: GO:0000028\n",
      "name: ribosomal small subunit assembly\n",
      "namespace: biological_process\n",
      "def: \"The aggregation, arrangement and bonding together of constituent RNAs and proteins to form the small ribosomal subunit.\" [GOC:jl, PMID:30467428]\n",
      "synonym: \"30S ribosomal subunit assembly\" NARROW [GOC:mah]\n",
      "synonym: \"40S ribosomal subunit assembly\" NARROW [GOC:mah]\n",
      "is_a: GO:0022618 ! ribonucleoprotein complex assembly\n",
      "relationship: part_of GO:0042255 ! ribosome assembly\n",
      "relationship: part_of GO:0042274 ! ribosomal small subunit biogenesis\n",
      "\n",
      "[Term]\n",
      "id: GO:0000030\n",
      "name: mannosyltransferase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the transfer of a mannosyl group to an acceptor molecule, typically another carbohydrate or a lipid.\" [GOC:ai, GOC:cjm]\n",
      "xref: Reactome:R-HSA-162797 \"mannose (a1-2) mannose (a1-6) (ethanolamineP) mannose (a1-4) glucosaminyl-acyl-PI -> mannose (a1) mannose (a1-2) mannose (a1-6) (ethanolamineP) mannose (a1-4) glucosaminyl-acyl-PI\"\n",
      "xref: Reactome:R-HSA-162830 \"glucosaminyl-acyl-PI + dolichol phosphate D-mannose -> mannose(al1-4)glucosaminyl-acyl-PI + dolichol phosphate\"\n",
      "xref: Reactome:R-HSA-446198 \"ALG12 transfers Man to N-glycan precursor (GlcNAc)2 (Man)7 (PP-Dol)1\"\n",
      "xref: Reactome:R-HSA-4720497 \"Defective ALG12 does not add mannose to the N-glycan precursor\"\n",
      "is_a: GO:0016758 ! hexosyltransferase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000031\n",
      "name: mannosylphosphate transferase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the transfer of a mannosylphosphate group from one compound to another.\" [GOC:jl]\n",
      "is_a: GO:0016758 ! hexosyltransferase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000032\n",
      "name: cell wall mannoprotein biosynthetic process\n",
      "namespace: biological_process\n",
      "def: \"The chemical reactions and pathways resulting in the formation of cell wall mannoproteins, any cell wall protein that contains covalently bound mannose residues.\" [GOC:ai]\n",
      "synonym: \"cell wall mannoprotein anabolism\" EXACT []\n",
      "synonym: \"cell wall mannoprotein biosynthesis\" EXACT []\n",
      "synonym: \"cell wall mannoprotein formation\" EXACT []\n",
      "synonym: \"cell wall mannoprotein synthesis\" EXACT []\n",
      "is_a: GO:0006057 ! mannoprotein biosynthetic process\n",
      "is_a: GO:0031506 ! cell wall glycoprotein biosynthetic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000033\n",
      "name: alpha-1,3-mannosyltransferase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the transfer of a mannose residue to an oligosaccharide, forming an alpha-(1->3) linkage.\" [GOC:mcc, PMID:10521541]\n",
      "xref: Reactome:R-HSA-446188 \"ALG3 transfers Man to N-glycan precursor (GlcNAc)2 (Man)5 (PP-Dol)1\"\n",
      "xref: Reactome:R-HSA-446208 \"Addition of a second mannose to the N-glycan precursor by ALG2\"\n",
      "xref: Reactome:R-HSA-4549368 \"Defective ALG2 does not transfer a second Man to N-glycan precursor\"\n",
      "xref: Reactome:R-HSA-4720473 \"Defective ALG3 does not add mannose to the N-glycan precursor\"\n",
      "is_a: GO:0000030 ! mannosyltransferase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000034\n",
      "name: adenine deaminase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the reaction: adenine + H2O = hypoxanthine + NH3.\" [EC:3.5.4.2]\n",
      "synonym: \"ADase activity\" RELATED [EC:3.5.4.2]\n",
      "synonym: \"adenase activity\" RELATED [EC:3.5.4.2]\n",
      "synonym: \"adenine aminase activity\" RELATED [EC:3.5.4.2]\n",
      "synonym: \"adenine aminohydrolase activity\" RELATED [EC:3.5.4.2]\n",
      "xref: EC:3.5.4.2\n",
      "xref: MetaCyc:ADENINE-DEAMINASE-RXN\n",
      "xref: RHEA:23688\n",
      "is_a: GO:0016814 ! hydrolase activity, acting on carbon-nitrogen (but not peptide) bonds, in cyclic amidines\n",
      "is_a: GO:0019239 ! deaminase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000035\n",
      "name: acyl binding\n",
      "namespace: molecular_function\n",
      "def: \"Binding to an acyl group, any group formally derived by removal of the hydroxyl group from the acid function of a carboxylic acid.\" [GOC:curators, ISBN:0198506732]\n",
      "synonym: \"acyl-CoA or acyl binding\" BROAD []\n",
      "is_a: GO:0005488 ! binding\n",
      "\n",
      "[Term]\n",
      "id: GO:0000036\n",
      "name: acyl carrier activity\n",
      "namespace: molecular_function\n",
      "def: \"Binding an acyl group and presenting it for processing or offloading to a cognate enzyme. Covalently binds the acyl group via a phosphopantetheine prosthetic group and mediates protein-protein interactions with the enzyme conferring specificity. The acyl carrier protein (ACP) presents substrates to enzymes involved in fatty acid biosynthesis or in polyketide secondary metabolite biosynthesis.\" [GOC:jl, GOC:vw]\n",
      "synonym: \"ACP phosphopantetheine attachment site binding involved in fatty acid biosynthetic process\" EXACT []\n",
      "is_a: GO:0044620 ! ACP phosphopantetheine attachment site binding\n",
      "is_a: GO:0140414 ! phosphopantetheine-dependent carrier activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000038\n",
      "name: very long-chain fatty acid metabolic process\n",
      "namespace: biological_process\n",
      "def: \"The chemical reactions and pathways involving a fatty acid which has a chain length greater than C22.\" [GOC:hjd]\n",
      "synonym: \"very long chain fatty acid metabolic process\" EXACT [GOC:bf]\n",
      "synonym: \"very-long-chain fatty acid metabolic process\" EXACT []\n",
      "synonym: \"very-long-chain fatty acid metabolism\" EXACT []\n",
      "is_a: GO:0006631 ! fatty acid metabolic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000039\n",
      "name: obsolete plasma membrane long-chain fatty acid transporter\n",
      "namespace: molecular_function\n",
      "def: \"OBSOLETE. (Was not defined before being made obsolete).\" [GOC:ai]\n",
      "comment: This term was made obsolete because it describes a gene product and it contains component information.\n",
      "synonym: \"plasma membrane long-chain fatty acid transporter\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0005324\n",
      "consider: GO:0005886\n",
      "\n",
      "[Term]\n",
      "id: GO:0000041\n",
      "name: transition metal ion transport\n",
      "namespace: biological_process\n",
      "def: \"The directed movement of transition metal ions into, out of or within a cell, or between cells, by means of some agent such as a transporter or pore. A transition metal is an element whose atom has an incomplete d-subshell of extranuclear electrons, or which gives rise to a cation or cations with an incomplete d-subshell. Transition metals often have more than one valency state. Biologically relevant transition metals include vanadium, manganese, iron, copper, cobalt, nickel, molybdenum and silver.\" [ISBN:0198506732]\n",
      "synonym: \"transition metal transport\" EXACT []\n",
      "is_a: GO:0030001 ! metal ion transport\n",
      "\n",
      "[Term]\n",
      "id: GO:0000044\n",
      "name: obsolete ascorbate stabilization\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. The reduction of the ascorbate free radical to a stable form.\" [GOC:ai, GOC:mtg_electron_transport]\n",
      "comment: This term was made obsolete because it is defined as a function term and is in the process ontology.\n",
      "synonym: \"ascorbate stabilization\" EXACT []\n",
      "synonym: \"vitamin C stabilization\" EXACT []\n",
      "is_obsolete: true\n",
      "\n",
      "[Term]\n",
      "id: GO:0000045\n",
      "name: autophagosome assembly\n",
      "namespace: biological_process\n",
      "def: \"The formation of a double membrane-bounded structure, the autophagosome, that occurs when a specialized membrane sac, called the isolation membrane, starts to enclose a portion of the cytoplasm.\" [GOC:autophagy, PMID:9412464]\n",
      "synonym: \"autophagic vacuole assembly\" EXACT [GOC:autophagy]\n",
      "synonym: \"autophagic vacuole formation\" RELATED [GOC:mah]\n",
      "synonym: \"autophagosome biosynthesis\" EXACT []\n",
      "synonym: \"autophagosome formation\" EXACT []\n",
      "synonym: \"PAS formation\" NARROW []\n",
      "is_a: GO:0070925 ! organelle assembly\n",
      "is_a: GO:1905037 ! autophagosome organization\n",
      "\n",
      "[Term]\n",
      "id: GO:0000047\n",
      "name: obsolete Rieske iron-sulfur protein\n",
      "namespace: molecular_function\n",
      "def: \"OBSOLETE. (Was not defined before being made obsolete).\" [GOC:ai]\n",
      "comment: This term was made obsolete because it represents a gene product.\n",
      "synonym: \"Rieske iron-sulfur protein\" EXACT []\n",
      "synonym: \"Rieske iron-sulphur protein\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0009055\n",
      "\n",
      "[Term]\n",
      "id: GO:0000048\n",
      "name: peptidyltransferase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the reaction: peptidyl-tRNA(1) + aminoacyl-tRNA(2) = tRNA(1) + peptidylaminoacyl-tRNA(2).\" [PMID:11433365, PMID:9242921]\n",
      "synonym: \"peptidyl-tRNA:aminoacyl-tRNA N-peptidyltransferase activity\" RELATED []\n",
      "xref: EC:2.3.2.12\n",
      "xref: MetaCyc:PEPTIDYLTRANSFERASE-RXN\n",
      "xref: Reactome:R-HSA-156912 \"Peptide transfer from P-site tRNA to the A-site tRNA\"\n",
      "is_a: GO:0016755 ! aminoacyltransferase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000049\n",
      "name: tRNA binding\n",
      "namespace: molecular_function\n",
      "alt_id: GO:0000946\n",
      "def: \"Binding to a transfer RNA.\" [GOC:ai]\n",
      "synonym: \"base pairing with tRNA\" NARROW []\n",
      "is_a: GO:0003723 ! RNA binding\n",
      "\n",
      "[Term]\n",
      "id: GO:0000050\n",
      "name: urea cycle\n",
      "namespace: biological_process\n",
      "alt_id: GO:0006594\n",
      "alt_id: GO:0006871\n",
      "def: \"The sequence of reactions by which arginine is synthesized from ornithine, then cleaved to yield urea and regenerate ornithine. The overall reaction equation is NH3 + CO2 + aspartate + 3 ATP + 2 H2O = urea + fumarate + 2 ADP + 2 phosphate + AMP + diphosphate.\" [GOC:pde, GOC:vw, ISBN:0198506732]\n",
      "synonym: \"ornithine cycle\" EXACT []\n",
      "synonym: \"urea biosynthesis\" EXACT []\n",
      "synonym: \"urea biosynthetic process\" EXACT []\n",
      "xref: Wikipedia:Urea_cycle\n",
      "is_a: GO:0019627 ! urea metabolic process\n",
      "is_a: GO:0043604 ! amide biosynthetic process\n",
      "is_a: GO:1901566 ! organonitrogen compound biosynthetic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000051\n",
      "name: obsolete urea cycle intermediate metabolic process\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. The chemical reactions and pathways involving any of the intermediate compounds involved in the urea cycle, a cyclic metabolic pathway that converts waste nitrogen in the form of ammonium to urea.\" [GOC:jl, ISBN:0198506732]\n",
      "comment: This term was made obsolete because it is a grouping term that is not useful, but has caused true path violations.\n",
      "synonym: \"urea cycle intermediate metabolic process\" EXACT []\n",
      "synonym: \"urea cycle intermediate metabolism\" EXACT []\n",
      "is_obsolete: true\n",
      "\n",
      "[Term]\n",
      "id: GO:0000052\n",
      "name: citrulline metabolic process\n",
      "namespace: biological_process\n",
      "def: \"The chemical reactions and pathways involving citrulline, N5-carbamoyl-L-ornithine, an alpha amino acid not found in proteins.\" [ISBN:0198506732]\n",
      "synonym: \"citrulline metabolism\" EXACT []\n",
      "is_a: GO:1901605 ! alpha-amino acid metabolic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000053\n",
      "name: argininosuccinate metabolic process\n",
      "namespace: biological_process\n",
      "def: \"The chemical reactions and pathways involving argininosuccinate, 2-(N(omega)-arginino)succinate, an intermediate in the ornithine-urea cycle, where it is synthesized from citrulline and aspartate.\" [ISBN:0198506732]\n",
      "synonym: \"argininosuccinate metabolism\" EXACT []\n",
      "is_a: GO:0006575 ! cellular modified amino acid metabolic process\n",
      "is_a: GO:0072350 ! tricarboxylic acid metabolic process\n",
      "is_a: GO:1901605 ! alpha-amino acid metabolic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000054\n",
      "name: ribosomal subunit export from nucleus\n",
      "namespace: biological_process\n",
      "def: \"The directed movement of a ribosomal subunit from the nucleus into the cytoplasm.\" [GOC:ai]\n",
      "subset: goslim_yeast\n",
      "synonym: \"ribosomal subunit export from cell nucleus\" EXACT [GOC:mah]\n",
      "synonym: \"ribosomal subunit export out of nucleus\" EXACT [GOC:mah]\n",
      "synonym: \"ribosomal subunit transport from nucleus to cytoplasm\" EXACT [GOC:mah]\n",
      "synonym: \"ribosomal subunit-nucleus export\" EXACT [GOC:mah]\n",
      "synonym: \"ribosome export from nucleus\" RELATED [GOC:mah, GOC:rb]\n",
      "is_a: GO:0031503 ! protein-containing complex localization\n",
      "is_a: GO:0033750 ! ribosome localization\n",
      "is_a: GO:0051168 ! nuclear export\n",
      "is_a: GO:0051656 ! establishment of organelle localization\n",
      "relationship: part_of GO:0042254 ! ribosome biogenesis\n",
      "\n",
      "[Term]\n",
      "id: GO:0000055\n",
      "name: ribosomal large subunit export from nucleus\n",
      "namespace: biological_process\n",
      "alt_id: GO:0000057\n",
      "def: \"The directed movement of a ribosomal large subunit from the nucleus into the cytoplasm.\" [GOC:mah]\n",
      "synonym: \"50S ribosomal subunit export from nucleus\" NARROW [GOC:mah]\n",
      "synonym: \"60S ribosomal subunit export from nucleus\" NARROW [GOC:mah]\n",
      "synonym: \"ribosomal large subunit export from cell nucleus\" EXACT []\n",
      "synonym: \"ribosomal large subunit export out of nucleus\" EXACT []\n",
      "synonym: \"ribosomal large subunit transport from nucleus to cytoplasm\" EXACT []\n",
      "synonym: \"ribosomal large subunit-nucleus export\" EXACT []\n",
      "is_a: GO:0000054 ! ribosomal subunit export from nucleus\n",
      "\n",
      "[Term]\n",
      "id: GO:0000056\n",
      "name: ribosomal small subunit export from nucleus\n",
      "namespace: biological_process\n",
      "alt_id: GO:0000058\n",
      "def: \"The directed movement of a ribosomal small subunit from the nucleus into the cytoplasm.\" [GOC:mah]\n",
      "synonym: \"30S ribosomal subunit export from nucleus\" NARROW [GOC:mah]\n",
      "synonym: \"40S ribosomal subunit export from nucleus\" NARROW [GOC:mah]\n",
      "synonym: \"ribosomal small subunit export from cell nucleus\" EXACT []\n",
      "synonym: \"ribosomal small subunit export out of nucleus\" EXACT []\n",
      "synonym: \"ribosomal small subunit transport from nucleus to cytoplasm\" EXACT []\n",
      "synonym: \"ribosomal small subunit-nucleus export\" EXACT []\n",
      "is_a: GO:0000054 ! ribosomal subunit export from nucleus\n",
      "\n",
      "[Term]\n",
      "id: GO:0000059\n",
      "name: obsolete protein import into nucleus, docking\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. A protein complex assembly process that contributes to protein import into the nucleus, and that results in the association of a cargo protein, a carrier protein such as an importin alpha/beta heterodimer, and a nucleoporin located at the periphery of the nuclear pore complex.\" [GOC:isa_complete, GOC:mah, PMID:14570049, PMID:7878057, PMID:9126736]\n",
      "comment: This term was made obsolete because the transient assembly is better captured as a protein-protein association, if at all.\n",
      "synonym: \"protein docking during protein import into nucleus\" EXACT []\n",
      "synonym: \"protein docking during protein transport from cytoplasm to nucleus\" EXACT []\n",
      "synonym: \"protein docking during protein-nucleus import\" EXACT []\n",
      "synonym: \"protein transport from cytoplasm to nucleus, docking\" EXACT []\n",
      "synonym: \"protein-nucleus import, docking\" EXACT []\n",
      "is_obsolete: true\n",
      "\n",
      "[Term]\n",
      "id: GO:0000060\n",
      "name: obsolete protein import into nucleus, translocation\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. A protein transport process that contributes to protein import into the nucleus, and that results in the vectorial transfer of a cargo-carrier protein complex through the nuclear pore complex from the cytoplasmic side to the nucleoplasmic side of the nuclear envelope.\" [GOC:curators, ISBN:0198506732, PMID:14570049, PMID:9126736]\n",
      "comment: This term has been obsoleted because it represents a substep of the parent (GO:0006606 protein import into nucleus), has been incorrectly used.\n",
      "synonym: \"protein import into cell nucleus, translocation\" EXACT []\n",
      "synonym: \"protein translocation during protein import into nucleus\" EXACT []\n",
      "synonym: \"protein translocation during protein transport from cytoplasm to nucleus\" EXACT []\n",
      "synonym: \"protein translocation during protein-nucleus import\" EXACT []\n",
      "synonym: \"protein transport from cytoplasm to nucleus, translocation\" EXACT []\n",
      "synonym: \"protein-nucleus import, translocation\" EXACT []\n",
      "is_obsolete: true\n",
      "\n",
      "[Term]\n",
      "id: GO:0000061\n",
      "name: obsolete protein import into nucleus, substrate release\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. A protein complex disassembly process that contributes to protein import into the nucleus, and that results in the dissociation of the cargo protein and the carrier (such as an importin alpha/beta heterodimer) from each other and from the nuclear pore complex.\" [GOC:mah, PMID:14570049, PMID:9126736, PMID:9687515]\n",
      "comment: This term has been obsoleted because it represents a substep of the parent (GO:0006606 protein import into nucleus), has been incorrectly used.\n",
      "synonym: \"protein import into cell nucleus, substrate release\" EXACT []\n",
      "synonym: \"protein substrate release during protein import into nucleus\" EXACT []\n",
      "synonym: \"protein substrate release during protein transport from cytoplasm to nucleus\" EXACT []\n",
      "synonym: \"protein substrate release during protein-nucleus import\" EXACT []\n",
      "synonym: \"protein transport from cytoplasm to nucleus, substrate release\" EXACT []\n",
      "synonym: \"protein-nucleus import, substrate release\" EXACT []\n",
      "is_obsolete: true\n",
      "\n",
      "[Term]\n",
      "id: GO:0000062\n",
      "name: fatty-acyl-CoA binding\n",
      "namespace: molecular_function\n",
      "def: \"Binding to a fatty-acyl-CoA, any derivative of coenzyme A in which the sulfhydryl group is in thiolester linkage with a fatty acyl group.\" [GOC:jl, GOC:krc, ISBN:0198506732]\n",
      "synonym: \"fatty-acyl binding\" BROAD []\n",
      "synonym: \"fatty-acyl-coenzyme A binding\" EXACT []\n",
      "is_a: GO:0120227 ! acyl-CoA binding\n",
      "is_a: GO:1901567 ! fatty acid derivative binding\n",
      "\n",
      "[Term]\n",
      "id: GO:0000064\n",
      "name: L-ornithine transmembrane transporter activity\n",
      "namespace: molecular_function\n",
      "def: \"Enables the transfer of L-ornithine from one side of a membrane to the other. L-ornithine is 2,5-diaminopentanoic acid.\" [GOC:ai, GOC:mtg_transport, ISBN:0815340729]\n",
      "synonym: \"histidine/arginine/lysine/ornithine porter activity\" RELATED []\n",
      "synonym: \"L-ornithine transporter activity\" BROAD []\n",
      "xref: Reactome:R-HSA-70634 \"ornithine (cytosolic) + citrulline (mitochondrial) => ornithine (mitochondrial) + citrulline (cytosolic)\"\n",
      "is_a: GO:0015179 ! L-amino acid transmembrane transporter activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000067\n",
      "name: obsolete DNA replication and chromosome cycle\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. (Was not defined before being made obsolete).\" [GOC:ai]\n",
      "comment: This term was made obsolete because it has been superseded by more accurate terms to represent the biological processes occurring, and it is not clear that this term represents a useful entity.\n",
      "synonym: \"DNA replication and chromosome cycle\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0006260\n",
      "consider: GO:0007059\n",
      "consider: GO:0051276\n",
      "\n",
      "[Term]\n",
      "id: GO:0000070\n",
      "name: mitotic sister chromatid segregation\n",
      "namespace: biological_process\n",
      "alt_id: GO:0016359\n",
      "def: \"The cell cycle process in which replicated homologous chromosomes are organized and then physically separated and apportioned to two sets during the mitotic cell cycle. Each replicated chromosome, composed of two sister chromatids, aligns at the cell equator, paired with its homologous partner. One homolog of each morphologic type goes into each of the resulting chromosome sets.\" [GOC:ai, GOC:jl]\n",
      "subset: goslim_pombe\n",
      "synonym: \"mitotic chromosome segregation\" EXACT []\n",
      "synonym: \"mitotic sister-chromatid adhesion release\" NARROW []\n",
      "is_a: GO:0000819 ! sister chromatid segregation\n",
      "is_a: GO:1903047 ! mitotic cell cycle process\n",
      "relationship: part_of GO:0140014 ! mitotic nuclear division\n",
      "\n",
      "[Term]\n",
      "id: GO:0000072\n",
      "name: obsolete M phase specific microtubule process\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. A microtubule-based process that occurs only during M phase of the cell cycle.\" [GOC:mah]\n",
      "comment: This term was made obsolete because terms already exist for centrosome/spindle organization which would be more suitable for the existing annotations, and the phase could be captured as an annotation extension if necessary.\n",
      "synonym: \"M phase specific microtubule process\" EXACT []\n",
      "synonym: \"M-phase specific microtubule process\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0007017\n",
      "\n",
      "[Term]\n",
      "id: GO:0000073\n",
      "name: initial mitotic spindle pole body separation\n",
      "namespace: biological_process\n",
      "alt_id: GO:0030475\n",
      "def: \"The release of duplicated mitotic spindle pole bodies (SPBs) that begins with the nucleation of microtubules from each SPB within the nucleus, leading to V-shaped spindle microtubules. Interpolar microtubules that elongate from each pole are interconnected, forming overlapping microtubules. Capturing and antiparallel sliding apart of microtubules promotes the initial separation of the SPB.\" [GOC:sgd_curators, GOC:vw]\n",
      "is_a: GO:0110100 ! spindle pole body separation\n",
      "is_a: GO:1903047 ! mitotic cell cycle process\n",
      "relationship: part_of GO:0061804 ! mitotic spindle formation (spindle phase one)\n",
      "relationship: part_of GO:1905047 ! mitotic spindle pole body organization\n",
      "\n",
      "[Term]\n",
      "id: GO:0000075\n",
      "name: cell cycle checkpoint signaling\n",
      "namespace: biological_process\n",
      "alt_id: GO:0031576\n",
      "alt_id: GO:0071779\n",
      "alt_id: GO:0072395\n",
      "alt_id: GO:0072404\n",
      "alt_id: GO:0072407\n",
      "def: \"A signaling process that controls cell cycle progression by monitoring the integrity of specific cell cycle events. A cell cycle checkpoint begins with detection of deficiencies or defects and ends with signal transduction.\" [GOC:mtg_cell_cycle]\n",
      "comment: This term should not be used in direct manual annotation as it should always be possible to minimally designate mitotic or meiotic checkpoint, and usually to additionally specify the checkpoint (i.e mitotic spindle assembly checkpoint, mitotic DNA damage checkpoint etc). Note also that the effector processes are not part of the checkpoint but are positively regulated by the checkpoint signaling and should not be annotated here.\n",
      "subset: gocheck_do_not_annotate\n",
      "synonym: \"cell cycle checkpoint\" EXACT []\n",
      "synonym: \"G1/S checkpoint\" NARROW []\n",
      "synonym: \"G1/S transition checkpoint\" NARROW []\n",
      "synonym: \"G2/M checkpoint\" NARROW []\n",
      "synonym: \"G2/M transition checkpoint\" NARROW []\n",
      "synonym: \"signal transduction involved in cell cycle checkpoint\" EXACT []\n",
      "synonym: \"signal transduction involved in G2/M transition checkpoint\" EXACT []\n",
      "xref: Wikipedia:Cell_cycle_checkpoint\n",
      "is_a: GO:0035556 ! intracellular signal transduction\n",
      "is_a: GO:1901988 ! negative regulation of cell cycle phase transition\n",
      "\n",
      "[Term]\n",
      "id: GO:0000076\n",
      "name: DNA replication checkpoint signaling\n",
      "namespace: biological_process\n",
      "alt_id: GO:0072437\n",
      "def: \"A signal transduction process that contributes to a DNA replication checkpoint, that prevents the initiation of nuclear division until DNA replication is complete, thereby ensuring that progeny inherit a full complement of the genome.\" [GOC:curators, GOC:rn, PMID:11728327, PMID:12537518]\n",
      "synonym: \"DNA replication checkpoint\" EXACT []\n",
      "synonym: \"signal transduction involved in DNA replication checkpoint\" EXACT []\n",
      "is_a: GO:0031570 ! DNA integrity checkpoint signaling\n",
      "\n",
      "[Term]\n",
      "id: GO:0000077\n",
      "name: DNA damage checkpoint signaling\n",
      "namespace: biological_process\n",
      "alt_id: GO:0072422\n",
      "def: \"A signal transduction process that contributes to a DNA damage checkpoint.\" [GOC:mah]\n",
      "synonym: \"DNA damage checkpoint\" EXACT []\n",
      "synonym: \"DNA damage response, signal transduction resulting in cell cycle arrest\" NARROW []\n",
      "synonym: \"signal transduction involved in DNA damage checkpoint\" EXACT []\n",
      "xref: Wikipedia:DNA_damage_checkpoint\n",
      "xref: Wikipedia:Postreplication_checkpoint\n",
      "is_a: GO:0031570 ! DNA integrity checkpoint signaling\n",
      "is_a: GO:0042770 ! signal transduction in response to DNA damage\n",
      "\n",
      "[Term]\n",
      "id: GO:0000078\n",
      "name: obsolete cytokinesis after mitosis checkpoint\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. A mitotic cell cycle checkpoint that detects whether chromosome segregation is complete and negatively regulates cytokinesis following mitosis.\" [GOC:mtg_cell_cycle]\n",
      "comment: The reason this term was made obsolete is that the two terms cytokinesis checkpoint (GO:0031565) and cytokinesis after mitosis (GO:0000078) were conflated in meaning.\n",
      "synonym: \"cell morphogenesis checkpoint\" EXACT [GOC:dph, GOC:vw]\n",
      "synonym: \"cell shape checkpoint\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0044878\n",
      "consider: GO:0044879\n",
      "\n",
      "[Term]\n",
      "id: GO:0000079\n",
      "name: regulation of cyclin-dependent protein serine/threonine kinase activity\n",
      "namespace: biological_process\n",
      "def: \"Any process that modulates the frequency, rate or extent of cyclin-dependent protein serine/threonine kinase activity.\" [GOC:go_curators, GOC:pr]\n",
      "synonym: \"regulation of CDK activity\" EXACT []\n",
      "synonym: \"regulation of cyclin-dependent protein kinase activity\" BROAD []\n",
      "is_a: GO:0071900 ! regulation of protein serine/threonine kinase activity\n",
      "is_a: GO:1904029 ! regulation of cyclin-dependent protein kinase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000080\n",
      "name: mitotic G1 phase\n",
      "namespace: biological_process\n",
      "def: \"The cell cycle 'gap' phase which is the interval between the completion of DNA segregation by mitosis and the beginning of DNA synthesis.\" [GOC:mtg_cell_cycle]\n",
      "comment: Note that this term should not be used for direct annotation. If you are trying to make an annotation to x phase, it is likely that the correct annotation is 'regulation of x/y phase transition' or to a process which occurs during the reported phase (i.e mitotic DNA replication for mitotic S-phase). To capture the phase when a specific location or process is observed, the phase term can be used in an annotation extension (PMID:24885854) applied to a cellular component term (with the relation exists_during) or a biological process term (with the relation happens_during).\n",
      "subset: gocheck_do_not_annotate\n",
      "synonym: \"G1 phase of mitotic cell cycle\" EXACT []\n",
      "is_a: GO:0051318 ! G1 phase\n",
      "relationship: part_of GO:0051329 ! mitotic interphase\n",
      "\n",
      "[Term]\n",
      "id: GO:0000082\n",
      "name: G1/S transition of mitotic cell cycle\n",
      "namespace: biological_process\n",
      "def: \"The mitotic cell cycle transition by which a cell in G1 commits to S phase. The process begins with the build up of G1 cyclin-dependent kinase (G1 CDK), resulting in the activation of transcription of G1 cyclins. The process ends with the positive feedback of the G1 cyclins on the G1 CDK which commits the cell to S phase, in which DNA replication is initiated.\" [GOC:mtg_cell_cycle]\n",
      "is_a: GO:0044772 ! mitotic cell cycle phase transition\n",
      "is_a: GO:0044843 ! cell cycle G1/S phase transition\n",
      "\n",
      "[Term]\n",
      "id: GO:0000083\n",
      "name: regulation of transcription involved in G1/S transition of mitotic cell cycle\n",
      "namespace: biological_process\n",
      "def: \"Any process that regulates transcription such that the target genes are involved in the transition between G1 and S phase of the mitotic cell cycle.\" [GOC:mtg_cell_cycle]\n",
      "is_a: GO:0006355 ! regulation of DNA-templated transcription\n",
      "is_a: GO:1903047 ! mitotic cell cycle process\n",
      "relationship: part_of GO:0000082 ! G1/S transition of mitotic cell cycle\n",
      "\n",
      "[Term]\n",
      "id: GO:0000084\n",
      "name: mitotic S phase\n",
      "namespace: biological_process\n",
      "def: \"The cell cycle phase, following G1, during which DNA synthesis takes place as part of a mitotic cell cycle.\" [GOC:mtg_cell_cycle]\n",
      "comment: Note that this term should not be used for direct annotation. If you are trying to make an annotation to x phase, it is likely that the correct annotation is 'regulation of x/y phase transition' or to a process which occurs during the reported phase (i.e mitotic DNA replication for mitotic S-phase). To capture the phase when a specific location or process is observed, the phase term can be used in an annotation extension (PMID:24885854) applied to a cellular component term (with the relation exists_during) or a biological process term (with the relation happens_during).\n",
      "subset: gocheck_do_not_annotate\n",
      "synonym: \"S phase of mitotic cell cycle\" EXACT []\n",
      "synonym: \"S-phase of mitotic cell cycle\" EXACT []\n",
      "is_a: GO:0051320 ! S phase\n",
      "relationship: part_of GO:0051329 ! mitotic interphase\n",
      "\n",
      "[Term]\n",
      "id: GO:0000085\n",
      "name: mitotic G2 phase\n",
      "namespace: biological_process\n",
      "def: \"The cell cycle 'gap' phase which is the interval between the completion of DNA synthesis and the beginning of DNA segregation by mitosis.\" [GOC:mtg_cell_cycle]\n",
      "comment: Note that this term should not be used for direct annotation. If you are trying to make an annotation to x phase, it is likely that the correct annotation is 'regulation of x/y phase transition' or to a process which occurs during the reported phase (i.e mitotic DNA replication for mitotic S-phase). To capture the phase when a specific location or process is observed, the phase term can be used in an annotation extension (PMID:24885854) applied to a cellular component term (with the relation exists_during) or a biological process term (with the relation happens_during).\n",
      "subset: gocheck_do_not_annotate\n",
      "synonym: \"G2 phase of mitotic cell cycle\" EXACT []\n",
      "is_a: GO:0051319 ! G2 phase\n",
      "relationship: part_of GO:0051329 ! mitotic interphase\n",
      "\n",
      "[Term]\n",
      "id: GO:0000086\n",
      "name: G2/M transition of mitotic cell cycle\n",
      "namespace: biological_process\n",
      "def: \"The mitotic cell cycle transition by which a cell in G2 commits to M phase. The process begins when the kinase activity of M cyclin/CDK complex reaches a threshold high enough for the cell cycle to proceed. This is accomplished by activating a positive feedback loop that results in the accumulation of unphosphorylated and active M cyclin/CDK complex.\" [GOC:mtg_cell_cycle]\n",
      "synonym: \"mitotic G2/M transition\" EXACT []\n",
      "is_a: GO:0044772 ! mitotic cell cycle phase transition\n",
      "is_a: GO:0044839 ! cell cycle G2/M phase transition\n",
      "\n",
      "[Term]\n",
      "id: GO:0000087\n",
      "name: mitotic M phase\n",
      "namespace: biological_process\n",
      "def: \"A cell cycle phase during which nuclear division occurs, and which is comprises the phases: prophase, metaphase, anaphase and telophase and occurs as part of a mitotic cell cycle.\" [GOC:mtg_cell_cycle]\n",
      "comment: Note that this term should not be used for direct annotation. If you are trying to make an annotation to x phase, it is likely that the correct annotation is 'regulation of x/y phase transition' or to a process which occurs during the reported phase (i.e mitotic DNA replication for mitotic S-phase). To capture the phase when a specific location or process is observed, the phase term can be used in an annotation extension (PMID:24885854) applied to a cellular component term (with the relation exists_during) or a biological process term (with the relation happens_during).\n",
      "subset: gocheck_do_not_annotate\n",
      "synonym: \"M phase of mitotic cell cycle\" EXACT []\n",
      "synonym: \"M-phase of mitotic cell cycle\" EXACT []\n",
      "is_a: GO:0000279 ! M phase\n",
      "is_a: GO:0098763 ! mitotic cell cycle phase\n",
      "\n",
      "[Term]\n",
      "id: GO:0000088\n",
      "name: mitotic prophase\n",
      "namespace: biological_process\n",
      "def: \"The cell cycle phase which is the first stage of M phase of mitosis and during which chromosomes condense and the two daughter centrioles and their asters migrate toward the poles of the cell.\" [GOC:mtg_cell_cycle]\n",
      "comment: Note that this term should not be used for direct annotation. If you are trying to make an annotation to x phase, it is likely that the correct annotation is 'regulation of x/y phase transition' or to a process which occurs during the reported phase (i.e mitotic DNA replication for mitotic S-phase). To capture the phase when a specific location or process is observed, the phase term can be used in an annotation extension (PMID:24885854) applied to a cellular component term (with the relation exists_during) or a biological process term (with the relation happens_during).\n",
      "subset: gocheck_do_not_annotate\n",
      "is_a: GO:0051324 ! prophase\n",
      "relationship: part_of GO:0000087 ! mitotic M phase\n",
      "\n",
      "[Term]\n",
      "id: GO:0000089\n",
      "name: mitotic metaphase\n",
      "namespace: biological_process\n",
      "def: \"The cell cycle phase, following prophase, during which chromosomes become aligned on the equatorial plate of the cell as part of a mitotic cell cycle.\" [GOC:mtg_cell_cycle]\n",
      "comment: Note that this term should not be used for direct annotation. If you are trying to make an annotation to x phase, it is likely that the correct annotation is 'regulation of x/y phase transition' or to a process which occurs during the reported phase (i.e mitotic DNA replication for mitotic S-phase). To capture the phase when a specific location or process is observed, the phase term can be used in an annotation extension (PMID:24885854) applied to a cellular component term (with the relation exists_during) or a biological process term (with the relation happens_during).\n",
      "subset: gocheck_do_not_annotate\n",
      "is_a: GO:0051323 ! metaphase\n",
      "relationship: part_of GO:0000087 ! mitotic M phase\n",
      "\n",
      "[Term]\n",
      "id: GO:0000090\n",
      "name: mitotic anaphase\n",
      "namespace: biological_process\n",
      "def: \"The cell cycle phase during which chromosomes separate and migrate towards the poles of the spindle the as part of a mitotic cell cycle.\" [GOC:mtg_cell_cycle]\n",
      "comment: Note that this term should not be used for direct annotation. If you are trying to make an annotation to x phase, it is likely that the correct annotation is 'regulation of x/y phase transition' or to a process which occurs during the reported phase (i.e mitotic DNA replication for mitotic S-phase). To capture the phase when a specific location or process is observed, the phase term can be used in an annotation extension (PMID:24885854) applied to a cellular component term (with the relation exists_during) or a biological process term (with the relation happens_during).\n",
      "subset: gocheck_do_not_annotate\n",
      "is_a: GO:0051322 ! anaphase\n",
      "relationship: part_of GO:0000087 ! mitotic M phase\n",
      "\n",
      "[Term]\n",
      "id: GO:0000091\n",
      "name: mitotic anaphase A\n",
      "namespace: biological_process\n",
      "def: \"The cell cycle phase during which the kinetochore microtubules shorten as chromosomes move toward the spindle poles as part of mitosis.\" [GOC:mtg_cell_cycle]\n",
      "comment: Note that this term should not be used for direct annotation. If you are trying to make an annotation to x phase, it is likely that the correct annotation is 'regulation of x/y phase transition' or to a process which occurs during the reported phase (i.e mitotic DNA replication for mitotic S-phase). To capture the phase when a specific location or process is observed, the phase term can be used in an annotation extension (PMID:24885854) applied to a cellular component term (with the relation exists_during) or a biological process term (with the relation happens_during).\n",
      "subset: gocheck_do_not_annotate\n",
      "is_a: GO:0000090 ! mitotic anaphase\n",
      "\n",
      "[Term]\n",
      "id: GO:0000092\n",
      "name: mitotic anaphase B\n",
      "namespace: biological_process\n",
      "def: \"The cell cycle phase during which the polar microtubules elongate and the two poles of the spindle move farther apart as part of mitosis.\" [GOC:mtg_cell_cycle]\n",
      "comment: Note that this term should not be used for direct annotation. If you are trying to make an annotation to x phase, it is likely that the correct annotation is 'regulation of x/y phase transition' or to a process which occurs during the reported phase (i.e mitotic DNA replication for mitotic S-phase). To capture the phase when a specific location or process is observed, the phase term can be used in an annotation extension (PMID:24885854) applied to a cellular component term (with the relation exists_during) or a biological process term (with the relation happens_during).\n",
      "subset: gocheck_do_not_annotate\n",
      "is_a: GO:0000090 ! mitotic anaphase\n",
      "\n",
      "[Term]\n",
      "id: GO:0000093\n",
      "name: mitotic telophase\n",
      "namespace: biological_process\n",
      "def: \"The cell cycle phase which follows anaphase during M phase of mitosis and during which the chromosomes arrive at the poles of the cell and the division of the cytoplasm starts.\" [GOC:mtg_cell_cycle]\n",
      "comment: Note that this term should not be used for direct annotation. If you are trying to make an annotation to x phase, it is likely that the correct annotation is 'regulation of x/y phase transition' or to a process which occurs during the reported phase (i.e mitotic DNA replication for mitotic S-phase). To capture the phase when a specific location or process is observed, the phase term can be used in an annotation extension (PMID:24885854) applied to a cellular component term (with the relation exists_during) or a biological process term (with the relation happens_during).\n",
      "subset: gocheck_do_not_annotate\n",
      "is_a: GO:0051326 ! telophase\n",
      "relationship: part_of GO:0000087 ! mitotic M phase\n",
      "\n",
      "[Term]\n",
      "id: GO:0000094\n",
      "name: obsolete septin assembly and septum formation\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. (Was not defined before being made obsolete).\" [GOC:go_curators]\n",
      "comment: This term was made obsolete because it was not defined and the string name implied two separate processes.\n",
      "synonym: \"septin assembly and septum formation\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0000917\n",
      "consider: GO:0000918\n",
      "consider: GO:0000921\n",
      "\n",
      "[Term]\n",
      "id: GO:0000095\n",
      "name: S-adenosyl-L-methionine transmembrane transporter activity\n",
      "namespace: molecular_function\n",
      "alt_id: GO:0015177\n",
      "def: \"Enables the transfer of S-adenosylmethionine from one side of a membrane to the other. S-adenosylmethionine is S-(5'-adenosyl)-L-methionine, an important intermediate in one-carbon metabolism.\" [GOC:ai]\n",
      "synonym: \"S-adenosyl methionine permease activity\" EXACT []\n",
      "synonym: \"S-adenosyl methionine transporter activity\" EXACT []\n",
      "synonym: \"S-adenosylmethionine permease activity\" EXACT []\n",
      "synonym: \"S-adenosylmethionine transmembrane transporter activity\" EXACT []\n",
      "synonym: \"S-adenosylmethionine transporter activity\" BROAD []\n",
      "synonym: \"SAM transmembrane transporter activity\" EXACT []\n",
      "xref: Reactome:R-HSA-8855062 \"SLC25A26 exchanges cytosolic AdoMet for mitochondrial AdoHcy\"\n",
      "is_a: GO:0015101 ! organic cation transmembrane transporter activity\n",
      "is_a: GO:1901682 ! sulfur compound transmembrane transporter activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000096\n",
      "name: sulfur amino acid metabolic process\n",
      "namespace: biological_process\n",
      "def: \"The chemical reactions and pathways involving amino acids containing sulfur, comprising cysteine, homocysteine, methionine and selenocysteine.\" [GOC:ai]\n",
      "synonym: \"sulfur amino acid metabolism\" EXACT []\n",
      "synonym: \"sulphur amino acid metabolic process\" EXACT []\n",
      "synonym: \"sulphur amino acid metabolism\" EXACT []\n",
      "is_a: GO:0006790 ! sulfur compound metabolic process\n",
      "is_a: GO:0019752 ! carboxylic acid metabolic process\n",
      "is_a: GO:1901564 ! organonitrogen compound metabolic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000097\n",
      "name: sulfur amino acid biosynthetic process\n",
      "namespace: biological_process\n",
      "def: \"The chemical reactions and pathways resulting in the formation of amino acids containing sulfur, comprising cysteine, methionine and selenocysteine.\" [GOC:ai]\n",
      "synonym: \"sulfur amino acid anabolism\" EXACT []\n",
      "synonym: \"sulfur amino acid biosynthesis\" EXACT []\n",
      "synonym: \"sulfur amino acid formation\" EXACT []\n",
      "synonym: \"sulfur amino acid synthesis\" EXACT []\n",
      "synonym: \"sulphur amino acid biosynthesis\" EXACT []\n",
      "synonym: \"sulphur amino acid biosynthetic process\" EXACT []\n",
      "is_a: GO:0000096 ! sulfur amino acid metabolic process\n",
      "is_a: GO:0044272 ! sulfur compound biosynthetic process\n",
      "is_a: GO:0046394 ! carboxylic acid biosynthetic process\n",
      "is_a: GO:1901566 ! organonitrogen compound biosynthetic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000098\n",
      "name: sulfur amino acid catabolic process\n",
      "namespace: biological_process\n",
      "def: \"The chemical reactions and pathways resulting in the breakdown of amino acids containing sulfur, comprising cysteine, methionine and selenocysteine.\" [GOC:ai]\n",
      "synonym: \"sulfur amino acid breakdown\" EXACT []\n",
      "synonym: \"sulfur amino acid catabolism\" EXACT []\n",
      "synonym: \"sulfur amino acid degradation\" EXACT []\n",
      "synonym: \"sulphur amino acid catabolic process\" EXACT []\n",
      "synonym: \"sulphur amino acid catabolism\" EXACT []\n",
      "is_a: GO:0000096 ! sulfur amino acid metabolic process\n",
      "is_a: GO:0044273 ! sulfur compound catabolic process\n",
      "is_a: GO:0046395 ! carboxylic acid catabolic process\n",
      "is_a: GO:1901565 ! organonitrogen compound catabolic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000099\n",
      "name: sulfur amino acid transmembrane transporter activity\n",
      "namespace: molecular_function\n",
      "def: \"Enables the transfer of sulfur amino acids from one side of a membrane to the other. Sulphur amino acids contain sulfur in the form of cystine, methionine or their derivatives.\" [GOC:ai, GOC:mtg_transport, ISBN:0815340729]\n",
      "synonym: \"sulfur amino acid transporter activity\" BROAD []\n",
      "synonym: \"sulphur amino acid transmembrane transporter activity\" EXACT []\n",
      "synonym: \"sulphur amino acid transporter activity\" BROAD []\n",
      "is_a: GO:0015171 ! amino acid transmembrane transporter activity\n",
      "is_a: GO:1901682 ! sulfur compound transmembrane transporter activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000100\n",
      "name: S-methylmethionine transmembrane transporter activity\n",
      "namespace: molecular_function\n",
      "alt_id: GO:0015178\n",
      "def: \"Enables the transfer of S-methylmethionine from one side of a membrane to the other.\" [GOC:ai]\n",
      "synonym: \"S-methylmethionine permease activity\" EXACT []\n",
      "synonym: \"S-methylmethionine transporter activity\" BROAD []\n",
      "is_a: GO:0072349 ! modified amino acid transmembrane transporter activity\n",
      "is_a: GO:1901682 ! sulfur compound transmembrane transporter activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000101\n",
      "name: sulfur amino acid transport\n",
      "namespace: biological_process\n",
      "def: \"The directed movement of amino acids containing sulfur (cystine, methionine and their derivatives) into, out of or within a cell, or between cells, by means of some agent such as a transporter or pore.\" [GOC:ai]\n",
      "synonym: \"sulphur amino acid transport\" EXACT []\n",
      "is_a: GO:0006865 ! amino acid transport\n",
      "is_a: GO:0072348 ! sulfur compound transport\n",
      "\n",
      "[Term]\n",
      "id: GO:0000102\n",
      "name: L-methionine secondary active transmembrane transporter activity\n",
      "namespace: molecular_function\n",
      "def: \"Enables the transfer of L-methionine from one side of a membrane to the other, up its concentration gradient. The transporter binds the solute and undergoes a series of conformational changes. Transport works equally well in either direction and is driven by a chemiosmotic source of energy. Secondary active transporters include symporters and antiporters.\" [GOC:mtg_transport]\n",
      "synonym: \"L-methionine porter activity\" RELATED []\n",
      "is_a: GO:0005294 ! neutral L-amino acid secondary active transmembrane transporter activity\n",
      "is_a: GO:0015191 ! L-methionine transmembrane transporter activity\n",
      "is_a: GO:1901680 ! sulfur-containing amino acid secondary active transmembrane transporter activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000103\n",
      "name: sulfate assimilation\n",
      "namespace: biological_process\n",
      "alt_id: GO:0019378\n",
      "def: \"The pathways by which inorganic sulfate is processed and incorporated into sulfated compounds.\" [GOC:jl]\n",
      "synonym: \"sulfate assimilation, phosphoadenylyl sulfate reduction by an oxidoreductase, acting on sulfur group of donors, NAD or NADP as acceptor\" NARROW []\n",
      "synonym: \"sulphate assimilation\" EXACT []\n",
      "synonym: \"sulphate assimilation, phosphoadenylyl sulphate reduction by an oxidoreductase, acting on sulphur group of donors, NAD or NADP as acceptor\" NARROW []\n",
      "is_a: GO:0006790 ! sulfur compound metabolic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000104\n",
      "name: succinate dehydrogenase activity\n",
      "namespace: molecular_function\n",
      "alt_id: GO:0019739\n",
      "def: \"Catalysis of the reaction: succinate + acceptor = fumarate + reduced acceptor.\" [GOC:kd]\n",
      "synonym: \"fumarate dehydrogenase activity\" EXACT []\n",
      "synonym: \"fumarate reductase activity\" EXACT []\n",
      "synonym: \"fumaric hydrogenase activity\" RELATED [EC:1.3.99.1]\n",
      "synonym: \"succinate oxidoreductase activity\" EXACT []\n",
      "synonym: \"succinate:(acceptor) oxidoreductase activity\" EXACT []\n",
      "synonym: \"succinate:acceptor oxidoreductase activity\" EXACT []\n",
      "synonym: \"succinic acid dehydrogenase activity\" EXACT []\n",
      "synonym: \"succinic dehydrogenase activity\" BROAD []\n",
      "synonym: \"succinodehydrogenase activity\" EXACT []\n",
      "synonym: \"succinyl dehydrogenase activity\" EXACT []\n",
      "xref: MetaCyc:SUCC-FUM-OXRED-RXN\n",
      "xref: RHEA:16357\n",
      "is_a: GO:0016627 ! oxidoreductase activity, acting on the CH-CH group of donors\n",
      "\n",
      "[Term]\n",
      "id: GO:0000105\n",
      "name: histidine biosynthetic process\n",
      "namespace: biological_process\n",
      "def: \"The chemical reactions and pathways resulting in the formation of histidine, 2-amino-3-(1H-imidazol-4-yl)propanoic acid.\" [GOC:go_curators]\n",
      "synonym: \"histidine anabolism\" EXACT []\n",
      "synonym: \"histidine biosynthesis\" EXACT []\n",
      "synonym: \"histidine formation\" EXACT []\n",
      "synonym: \"histidine synthesis\" EXACT []\n",
      "xref: MetaCyc:HISTSYN-PWY\n",
      "is_a: GO:0006547 ! histidine metabolic process\n",
      "is_a: GO:0008652 ! amino acid biosynthetic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000107\n",
      "name: imidazoleglycerol-phosphate synthase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the reaction: phosphoribulosylformimino-AICAR-P + L-glutamine = D-erythro-imidazole-glycerol-phosphate + aminoimidazole carboxamide ribonucleotide + L-glutamate + 2 H+.\" [RHEA:24793]\n",
      "synonym: \"glutamine amidotransferase:cyclase activity\" BROAD []\n",
      "synonym: \"imidazole glycerol phosphate synthase activity\" EXACT []\n",
      "synonym: \"imidazole-glycerol-phosphate synthase activity\" RELATED []\n",
      "synonym: \"imidazoleglycerol phosphate synthase activity\" EXACT []\n",
      "xref: MetaCyc:GLUTAMIDOTRANS-RXN\n",
      "xref: RHEA:24793\n",
      "is_a: GO:0016763 ! pentosyltransferase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000108\n",
      "name: obsolete repairosome\n",
      "namespace: cellular_component\n",
      "def: \"OBSOLETE. A stable complex of proteins that carry out the DNA damage recognition and incision reactions characteristic of nucleotide excision repair (NER), such as DNA damage recognition, DNA helix unwinding, and endonucleolytic cleavage at sites flanking damaged DNA; includes TFIIH subunits and additional polypeptides; may form in the absence of DNA damage.\" [PMID:10681587, PMID:9852079]\n",
      "comment: This term was made obsolete because 'repairosome' has fallen out of use in the literature, and the large complex described in the definition has not been confirmed to exist. The term has also confused annotators.\n",
      "synonym: \"repairosome\" EXACT []\n",
      "is_obsolete: true\n",
      "replaced_by: GO:0000109\n",
      "\n",
      "[Term]\n",
      "id: GO:0000109\n",
      "name: nucleotide-excision repair complex\n",
      "namespace: cellular_component\n",
      "def: \"Any complex formed of proteins that act in nucleotide-excision repair.\" [PMID:10915862]\n",
      "comment: Note that process information is included in the term and definition for the purpose of describing and distinguishing the complex.\n",
      "subset: goslim_pir\n",
      "synonym: \"UvrB-UvrC complex\" NARROW [PMID:12145219]\n",
      "synonym: \"UvrBC complex\" NARROW [GOC:bhm, PMID:12145219]\n",
      "is_a: GO:0140513 ! nuclear protein-containing complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000110\n",
      "name: nucleotide-excision repair factor 1 complex\n",
      "namespace: cellular_component\n",
      "def: \"One of several protein complexes involved in nucleotide-excision repair; possesses DNA damage recognition and endodeoxynuclease activities. In S. cerevisiae, it is composed of Rad1p, Rad10p, and Rad14p; in human the subunits are ERCC4/XPF, ERCC1 and XPA, respectively.\" [PMID:10915862]\n",
      "comment: Note that process and function information are included in the term and definition for the purpose of describing and distinguishing the complex.\n",
      "synonym: \"NEF1 complex\" EXACT []\n",
      "synonym: \"XPA-ERCC1-ERCC4 complex\" EXACT [PMID:8197175]\n",
      "is_a: GO:0000109 ! nucleotide-excision repair complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000111\n",
      "name: nucleotide-excision repair factor 2 complex\n",
      "namespace: cellular_component\n",
      "def: \"One of several protein complexes involved in nucleotide-excision repair; possesses damaged DNA binding activity. In S. cerevisiae, it is composed of Rad4p and Rad23p.\" [PMID:10915862]\n",
      "comment: Note that process and function information are included in the term and definition for the purpose of describing and distinguishing the complex.\n",
      "synonym: \"NEF2 complex\" EXACT []\n",
      "is_a: GO:0000109 ! nucleotide-excision repair complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000112\n",
      "name: nucleotide-excision repair factor 3 complex\n",
      "namespace: cellular_component\n",
      "def: \"One of several protein complexes involved in nucleotide-excision repair; possesses endodeoxynuclease and DNA helicase activities. In S. cerevisiae, it is composed of Rad2p and the core TFIIH-Ssl2p complex (core TFIIH is composed of Rad3p, Tfb1p, Tfb2p, Ssl1p, Tfb4p and Tfb5p. Note that Ssl2p is also called Rad25p).\" [GOC:ew, PMID:10915862, PMID:14500720, PMID:7813015]\n",
      "comment: Note that process and function information are included in the term and definition for the purpose of describing and distinguishing the complex.\n",
      "synonym: \"NEF3 complex\" EXACT []\n",
      "is_a: GO:0000109 ! nucleotide-excision repair complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000113\n",
      "name: nucleotide-excision repair factor 4 complex\n",
      "namespace: cellular_component\n",
      "def: \"One of several protein complexes involved in nucleotide-excision repair; possesses DNA damage recognition and DNA-dependent ATPase activities. In S. cerevisiae, it is composed of Rad7p and Rad16p.\" [PMID:10915862]\n",
      "comment: Note that process and function information are included in the term and definition for the purpose of describing and distinguishing the complex.\n",
      "synonym: \"NEF4 complex\" EXACT []\n",
      "is_a: GO:0000109 ! nucleotide-excision repair complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000114\n",
      "name: obsolete regulation of transcription involved in G1 phase of mitotic cell cycle\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. Any process that regulates transcription such that the target genes are transcribed as part of the G1 phase of the mitotic cell cycle.\" [GOC:dph, GOC:mah, GOC:tb]\n",
      "comment: This term was made obsolete because it is unclear exactly what it means. It could mean either 'regulation of transcription during phase X' or 'regulation of transition between phase X and phase Y'.\n",
      "synonym: \"G1-specific transcription in mitotic cell cycle\" RELATED []\n",
      "synonym: \"regulation of transcription from RNA polymerase II promoter during G1 phase of cell cycle\" EXACT []\n",
      "synonym: \"regulation of transcription involved in G1 phase of mitotic cell cycle\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0000083\n",
      "consider: GO:0006357\n",
      "\n",
      "[Term]\n",
      "id: GO:0000115\n",
      "name: obsolete regulation of transcription involved in S phase of mitotic cell cycle\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. A cell cycle process that regulates transcription such that the target genes are transcribed as part of the S phase of the mitotic cell cycle.\" [GOC:dph, GOC:mah, GOC:tb]\n",
      "comment: This term was made obsolete because it is unclear exactly what it means. It could mean either 'regulation of transcription during phase X' or 'regulation of transition between phase X and phase Y'.\n",
      "synonym: \"regulation of transcription from RNA polymerase II promoter during S-phase of mitotic cell cycle\" RELATED []\n",
      "synonym: \"regulation of transcription involved in S phase of mitotic cell cycle\" EXACT []\n",
      "synonym: \"regulation of transcription involved in S-phase of mitotic cell cycle\" EXACT []\n",
      "synonym: \"S-phase-specific transcription in mitotic cell cycle\" RELATED []\n",
      "synonym: \"S-specific transcription in mitotic cell cycle\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0000083\n",
      "consider: GO:0006357\n",
      "\n",
      "[Term]\n",
      "id: GO:0000116\n",
      "name: obsolete regulation of transcription involved in G2-phase of mitotic cell cycle\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. Any process that regulates transcription such that the target genes are transcribed as part of the G2 phase of the mitotic cell cycle.\" [GOC:dph, GOC:mah, GOC:tb]\n",
      "comment: This term was made obsolete because it is unclear exactly what it means. It could mean either 'regulation of transcription during phase X' or 'regulation of transition between phase X and phase Y'.\n",
      "synonym: \"G2-specific transcription in mitotic cell cycle\" RELATED []\n",
      "synonym: \"regulation of transcription from RNA polymerase II during G2-phase of mitotic cell cycle\" RELATED []\n",
      "synonym: \"regulation of transcription involved in G2-phase of mitotic cell cycle\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0000117\n",
      "consider: GO:0006357\n",
      "\n",
      "[Term]\n",
      "id: GO:0000117\n",
      "name: regulation of transcription involved in G2/M transition of mitotic cell cycle\n",
      "namespace: biological_process\n",
      "def: \"Any process that regulates transcription such that the target genes are transcribed as part of the G2/M transition of the mitotic cell cycle.\" [GOC:mtg_cell_cycle]\n",
      "synonym: \"G2/M-specific transcription in mitotic cell cycle\" RELATED []\n",
      "synonym: \"regulation of transcription from RNA polymerase II promoter during G2/M transition of mitotic cell cycle\" RELATED []\n",
      "is_a: GO:0006355 ! regulation of DNA-templated transcription\n",
      "is_a: GO:1903047 ! mitotic cell cycle process\n",
      "relationship: part_of GO:0000086 ! G2/M transition of mitotic cell cycle\n",
      "\n",
      "[Term]\n",
      "id: GO:0000118\n",
      "name: histone deacetylase complex\n",
      "namespace: cellular_component\n",
      "def: \"A protein complex that possesses histone deacetylase activity.\" [GOC:mah]\n",
      "comment: Note that this term represents a location, not a function; the activity possessed by this complex is mentioned in the definition for the purpose of describing and distinguishing the complex. The function of this complex is represented by the molecular function term 'histone deacetylase activity ; GO:0004407'.\n",
      "synonym: \"HDAC complex\" EXACT []\n",
      "is_a: GO:0140513 ! nuclear protein-containing complex\n",
      "is_a: GO:1902494 ! catalytic complex\n",
      "relationship: part_of GO:0005654 ! nucleoplasm\n",
      "\n",
      "[Term]\n",
      "id: GO:0000120\n",
      "name: RNA polymerase I transcription regulator complex\n",
      "namespace: cellular_component\n",
      "def: \"A transcription factor complex that acts at a regulatory region of a gene transcribed by RNA polymerase I.\" [GOC:mah]\n",
      "synonym: \"RNA polymerase I transcription factor complex\" NARROW []\n",
      "is_a: GO:0005667 ! transcription regulator complex\n",
      "is_a: GO:0140513 ! nuclear protein-containing complex\n",
      "relationship: part_of GO:0005730 ! nucleolus\n",
      "\n",
      "[Term]\n",
      "id: GO:0000121\n",
      "name: glycerol-1-phosphatase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the reaction: glycerol-1-phosphate + H2O = glycerol + phosphate.\" [EC:3.1.3.21]\n",
      "synonym: \"alpha-glycerol phosphatase activity\" RELATED [EC:3.1.3.21]\n",
      "synonym: \"alpha-glycerophosphatase activity\" RELATED [EC:3.1.3.21]\n",
      "synonym: \"glycerol 3-phosphatase activity\" RELATED [EC:3.1.3.21]\n",
      "synonym: \"glycerol 3-phosphate phosphohydrolase activity\" RELATED [EC:3.1.3.21]\n",
      "synonym: \"glycerol-1-phosphate phosphohydrolase activity\" RELATED [EC:3.1.3.21]\n",
      "synonym: \"glycerol-3-phosphate phosphatase activity\" RELATED [EC:3.1.3.21]\n",
      "xref: EC:3.1.3.21\n",
      "xref: MetaCyc:GLYCEROL-1-PHOSPHATASE-RXN\n",
      "xref: RHEA:11476\n",
      "is_a: GO:0016791 ! phosphatase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000122\n",
      "name: negative regulation of transcription by RNA polymerase II\n",
      "namespace: biological_process\n",
      "alt_id: GO:0010553\n",
      "alt_id: GO:0045816\n",
      "def: \"Any process that stops, prevents, or reduces the frequency, rate or extent of transcription mediated by RNA polymerase II.\" [GOC:go_curators, GOC:txnOH]\n",
      "synonym: \"down regulation of global transcription from RNA polymerase II promoter\" RELATED []\n",
      "synonym: \"down regulation of transcription from RNA polymerase II promoter\" EXACT []\n",
      "synonym: \"down-regulation of global transcription from RNA polymerase II promoter\" RELATED []\n",
      "synonym: \"down-regulation of transcription from RNA polymerase II promoter\" EXACT []\n",
      "synonym: \"downregulation of global transcription from RNA polymerase II promoter\" RELATED []\n",
      "synonym: \"downregulation of transcription from RNA polymerase II promoter\" EXACT []\n",
      "synonym: \"inhibition of global transcription from RNA polymerase II promoter\" RELATED []\n",
      "synonym: \"inhibition of transcription from RNA polymerase II promoter\" EXACT []\n",
      "synonym: \"negative regulation of gene-specific transcription from RNA polymerase II promoter\" RELATED []\n",
      "synonym: \"negative regulation of global transcription from Pol II promoter\" RELATED []\n",
      "synonym: \"negative regulation of transcription from Pol II promoter\" EXACT []\n",
      "synonym: \"negative regulation of transcription from RNA polymerase II promoter\" EXACT []\n",
      "synonym: \"negative regulation of transcription from RNA polymerase II promoter, global\" RELATED []\n",
      "is_a: GO:0006357 ! regulation of transcription by RNA polymerase II\n",
      "is_a: GO:0045892 ! negative regulation of DNA-templated transcription\n",
      "relationship: negatively_regulates GO:0006366 ! transcription by RNA polymerase II\n",
      "\n",
      "[Term]\n",
      "id: GO:0000123\n",
      "name: histone acetyltransferase complex\n",
      "namespace: cellular_component\n",
      "def: \"A protein complex that possesses histone acetyltransferase activity.\" [GOC:mah]\n",
      "comment: Note that this term represents a protein complex, not a function; the activity possessed by this complex is mentioned in the definition for the purpose of describing and distinguishing the complex. The function of this complex is represented by the molecular function term 'histone acetyltransferase activity ; GO:0004402'.\n",
      "synonym: \"histone acetylase complex\" EXACT []\n",
      "is_a: GO:0031248 ! protein acetyltransferase complex\n",
      "relationship: part_of GO:0000785 ! chromatin\n",
      "\n",
      "[Term]\n",
      "id: GO:0000124\n",
      "name: SAGA complex\n",
      "namespace: cellular_component\n",
      "alt_id: GO:0000125\n",
      "alt_id: GO:0030914\n",
      "def: \"A SAGA-type histone acetyltransferase complex that deubiquitinates H2A and/or H2B. This complex is organized into several functional submodules: a structural core including the activator binding module and consisting of ADA1 or a homolog, members of the SPT and TAF protein families as well as promotor recruitment factor TRRAP/TRA1, a histone acetyltransferase (HAT) module consisting of GCN5/KAT2A or PCAF/KAT2B, ADA2, ADA3/NGG1, and SGF29 or homologues thereof, a histone deubiquitinase (DUB) module consisting of ATXN7/SGF73, ATXN7L3/SGF11, ENY2/SUS1 and USP22/UBP8 or homologues thereof, and in some taxa a splicing module consisting of SF3B3 and SF3B5 or homologues thereof (not in fungi). In budding yeast also contains Spt8 which distinguishes it from SAGA-like (SLIK) complex (GO:0046695).\" [PMID:10637607, PMID:17337012, PMID:19056896, PMID:20838651, PMID:33004486]\n",
      "synonym: \"PCAF complex\" NARROW []\n",
      "synonym: \"PCAF histone acetylase-associated complex\" NARROW []\n",
      "synonym: \"Spt-Ada-Gcn5-acetyltransferase complex\" EXACT []\n",
      "synonym: \"SPT3-TAF9-GCN5 acetylase complex\" RELATED [GOC:rl, PMID:18838386]\n",
      "synonym: \"SPT3-TAF9-PCAF acetylase complex\" NARROW [PMID:18838386]\n",
      "synonym: \"STAGA coactivator complex\" RELATED []\n",
      "synonym: \"STAGA complex\" RELATED [PMID:32616828, PMID:33004486, PMID:34112237]\n",
      "is_a: GO:0070461 ! SAGA-type complex\n",
      "is_a: GO:1905368 ! peptidase complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000126\n",
      "name: transcription factor TFIIIB complex\n",
      "namespace: cellular_component\n",
      "def: \"A transcription factor complex that is involved in regulating transcription from RNA polymerase III (Pol III) promoters. TFIIIB contains the TATA-binding protein (TBP) and two Pol III-specific proteins, B'' and BRF.\" [GOC:mah, PMID:11433012]\n",
      "is_a: GO:0090576 ! RNA polymerase III transcription regulator complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000127\n",
      "name: transcription factor TFIIIC complex\n",
      "namespace: cellular_component\n",
      "def: \"A heterotrimeric transcription factor complex that is involved in regulating transcription from RNA polymerase III (Pol III) promoters. TFIIIC contains three conserved subunits that associate with the proximal Pol III promoter element, and additional subunits that associate with sequence elements downstream of the promoter and are more diverged among species. It also functions as a boundary element to partition genome content into distinct domains outside Pol III promoter regions.\" [GOC:mah, GOC:vw, PMID:11433012, PMID:16751097]\n",
      "is_a: GO:0090576 ! RNA polymerase III transcription regulator complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000128\n",
      "name: flocculation\n",
      "namespace: biological_process\n",
      "alt_id: GO:0000501\n",
      "alt_id: GO:0032128\n",
      "alt_id: GO:0036281\n",
      "alt_id: GO:0036282\n",
      "alt_id: GO:0043689\n",
      "alt_id: GO:0043690\n",
      "def: \"The reversible, non-sexual aggregation of single-celled organisms in suspension to form aggregates of many cells known as flocs.\" [GOC:jl, GOC:vw, PMID:11472912, PMID:21114594, PMID:8740415]\n",
      "comment: The word floc derives from the Latin word floccus, which means a tuft of wool.\n",
      "subset: goslim_pir\n",
      "synonym: \"cell-cell adhesion involved in flocculation\" EXACT []\n",
      "synonym: \"cell-cell adhesion involved in flocculation via cell wall protein-carbohydrate interaction\" NARROW []\n",
      "synonym: \"co-flocculation\" NARROW [PMID:10689163]\n",
      "synonym: \"coflocculation\" NARROW []\n",
      "synonym: \"coflocculation via lectin-mannose interaction\" NARROW [PMID:11693916]\n",
      "synonym: \"coflocculation via protein-carbohydrate interaction\" NARROW []\n",
      "synonym: \"flocculation via cell wall protein-carbohydrate interaction\" RELATED []\n",
      "synonym: \"flocculation via extracellular polymer\" RELATED []\n",
      "xref: Wikipedia:Flocculation\n",
      "is_a: GO:0051703 ! biological process involved in intraspecies interaction between organisms\n",
      "is_a: GO:0098610 ! adhesion between unicellular organisms\n",
      "\n",
      "[Term]\n",
      "id: GO:0000131\n",
      "name: incipient cellular bud site\n",
      "namespace: cellular_component\n",
      "def: \"The portion of the budding yeast plasma membrane where a daughter cell will emerge. The yeast marks this spot with bud-site selection proteins before bud emergence occurs. Actin is polarized to this spot just prior to and during bud emergence.\" [GOC:clt]\n",
      "is_a: GO:0140535 ! intracellular protein-containing complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000132\n",
      "name: establishment of mitotic spindle orientation\n",
      "namespace: biological_process\n",
      "alt_id: GO:0030607\n",
      "alt_id: GO:0030609\n",
      "def: \"A cell cycle process that sets the alignment of mitotic spindle relative to other cellular structures.\" [GOC:ems]\n",
      "synonym: \"establishment of spindle orientation during mitosis\" RELATED [GOC:dph, GOC:tb]\n",
      "synonym: \"establishment of spindle orientation involved in mitotic cell cycle\" EXACT [GOC:dph, GOC:tb]\n",
      "synonym: \"mitotic spindle orientation\" EXACT []\n",
      "synonym: \"orienting of mitotic spindle\" EXACT []\n",
      "is_a: GO:0040001 ! establishment of mitotic spindle localization\n",
      "is_a: GO:0051294 ! establishment of spindle orientation\n",
      "\n",
      "[Term]\n",
      "id: GO:0000133\n",
      "name: polarisome\n",
      "namespace: cellular_component\n",
      "def: \"Protein complex that plays a role in determining cell polarity by directing the localized assembly of actin filaments at polarization sites; in Saccharomyces the polarisome includes Bni1p, Spa2p, Pea2p, and Bud6p.\" [PMID:14734532, PMID:14998522, PMID:9632790]\n",
      "subset: goslim_pir\n",
      "is_a: GO:0032991 ! protein-containing complex\n",
      "relationship: part_of GO:0005938 ! cell cortex\n",
      "relationship: part_of GO:0030427 ! site of polarized growth\n",
      "\n",
      "[Term]\n",
      "id: GO:0000136\n",
      "name: mannan polymerase complex\n",
      "namespace: cellular_component\n",
      "def: \"A protein complex with alpha-(1->6)-mannosyltransferase activity, located in the cis Golgi membrane; adds mannan to N-linked glycans on proteins.\" [GOC:mcc, PMID:10037752, PMID:11095735, PMID:18083825]\n",
      "synonym: \"alpha-1,6-mannosyltransferase complex\" EXACT []\n",
      "is_a: GO:0031501 ! mannosyltransferase complex\n",
      "is_a: GO:0098796 ! membrane protein complex\n",
      "relationship: part_of GO:0000137 ! Golgi cis cisterna\n",
      "relationship: part_of GO:0000139 ! Golgi membrane\n",
      "\n",
      "[Term]\n",
      "id: GO:0000137\n",
      "name: Golgi cis cisterna\n",
      "namespace: cellular_component\n",
      "def: \"The Golgi cisterna closest to the endoplasmic reticulum; the first processing compartment through which proteins pass after export from the ER.\" [ISBN:0815316194]\n",
      "is_a: GO:0031985 ! Golgi cisterna\n",
      "\n",
      "[Term]\n",
      "id: GO:0000138\n",
      "name: Golgi trans cisterna\n",
      "namespace: cellular_component\n",
      "def: \"The Golgi cisterna farthest from the endoplasmic reticulum; the final processing compartment through which proteins pass before exiting the Golgi apparatus; the compartment in which N-linked protein glycosylation is completed.\" [ISBN:0815316194]\n",
      "synonym: \"late Golgi\" RELATED [GOC:mah]\n",
      "is_a: GO:0031985 ! Golgi cisterna\n",
      "\n",
      "[Term]\n",
      "id: GO:0000139\n",
      "name: Golgi membrane\n",
      "namespace: cellular_component\n",
      "def: \"The lipid bilayer surrounding any of the compartments of the Golgi apparatus.\" [GOC:mah]\n",
      "is_a: GO:0098588 ! bounding membrane of organelle\n",
      "relationship: part_of GO:0005794 ! Golgi apparatus\n",
      "\n",
      "[Term]\n",
      "id: GO:0000140\n",
      "name: acylglycerone-phosphate reductase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the reaction: 1-palmitoylglycerol-3-phosphate + NADP+ = palmitoylglycerone phosphate + NADPH + H+.\" [EC:1.1.1.101]\n",
      "synonym: \"1-acyldihydroxyacetone-phosphate reductase activity\" EXACT []\n",
      "synonym: \"1-palmitoylglycerol-3-phosphate:NADP+ oxidoreductase activity\" RELATED [EC:1.1.1.101]\n",
      "synonym: \"acyldihydroxyacetone phosphate reductase activity\" RELATED [EC:1.1.1.101]\n",
      "synonym: \"palmitoyl dihydroxyacetone phosphate reductase activity\" RELATED [EC:1.1.1.101]\n",
      "synonym: \"palmitoyl-dihydroxyacetone-phosphate reductase activity\" RELATED [EC:1.1.1.101]\n",
      "synonym: \"palmitoyldihydroxyacetone-phosphate reductase activity\" RELATED [EC:1.1.1.101]\n",
      "xref: EC:1.1.1.101\n",
      "xref: MetaCyc:ACYLGLYCERONE-PHOSPHATE-REDUCTASE-RXN\n",
      "xref: Reactome:R-HSA-75883 \"DHRS7B reduces GO3P to HXDG3P\"\n",
      "xref: RHEA:17341\n",
      "is_a: GO:0016616 ! oxidoreductase activity, acting on the CH-OH group of donors, NAD or NADP as acceptor\n",
      "\n",
      "[Term]\n",
      "id: GO:0000142\n",
      "name: cellular bud neck contractile ring\n",
      "namespace: cellular_component\n",
      "def: \"A contractile ring, i.e. a cytoskeletal structure composed of actin filaments and myosin, that forms beneath the plasma membrane at the mother-bud neck in mitotic cells that divide by budding in preparation for completing cytokinesis. An example of this structure is found in Saccharomyces cerevisiae.\" [GOC:krc, PMID:16009555]\n",
      "synonym: \"neck ring\" EXACT []\n",
      "is_a: GO:0110085 ! mitotic actomyosin contractile ring\n",
      "relationship: part_of GO:0005935 ! cellular bud neck\n",
      "\n",
      "[Term]\n",
      "id: GO:0000144\n",
      "name: cellular bud neck septin ring\n",
      "namespace: cellular_component\n",
      "def: \"A ring-shaped structure that forms at the site of cytokinesis in the bud neck of a budding cell; composed of members of the conserved family of filament forming proteins called septins as well as septin-associated proteins. In S. cerevisiae, this structure forms at the time of bud emergence and the septins show a high rate of exchange.\" [GOC:krc, PMID:16009555]\n",
      "is_a: GO:0000399 ! cellular bud neck septin structure\n",
      "is_a: GO:0005940 ! septin ring\n",
      "is_a: GO:0032161 ! cleavage apparatus septin structure\n",
      "\n",
      "[Term]\n",
      "id: GO:0000145\n",
      "name: exocyst\n",
      "namespace: cellular_component\n",
      "def: \"A protein complex peripherally associated with the plasma membrane that determines where vesicles dock and fuse. At least eight complex components are conserved between yeast and mammals.\" [GOC:cilia, PMID:15292201, PMID:27243008, PMID:9700152]\n",
      "synonym: \"exocyst complex\" EXACT []\n",
      "synonym: \"Sec6/8 complex\" EXACT []\n",
      "xref: Wikipedia:Exocyst\n",
      "is_a: GO:0099023 ! vesicle tethering complex\n",
      "relationship: part_of GO:0005938 ! cell cortex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000146\n",
      "name: microfilament motor activity\n",
      "namespace: molecular_function\n",
      "alt_id: GO:0030898\n",
      "def: \"A motor activity that generates movement along a microfilament, driven by ATP hydrolysis.\" [PMID:29716949]\n",
      "synonym: \"actin filament motor activity\" EXACT [GOC:dph]\n",
      "synonym: \"actin-activated ATPase activity\" EXACT []\n",
      "synonym: \"actin-dependent ATPase activity\" EXACT []\n",
      "synonym: \"actin-filament motor activity\" EXACT []\n",
      "synonym: \"muscle motor activity\" NARROW []\n",
      "synonym: \"myosin ATPase activity\" RELATED []\n",
      "xref: EC:5.6.1.8\n",
      "xref: Reactome:R-HSA-2316352 \"SLC2A4 (GLUT4) vesicle translocates and docks at the plasma membrane\"\n",
      "xref: Reactome:R-HSA-432237 \"Translocation of Aquaporin-2 from intracellular vesicles to the apical plasma membrane\"\n",
      "xref: Reactome:R-HSA-9023171 \"Insulin secretory granule translocates across the cortical actin network\"\n",
      "is_a: GO:0003774 ! cytoskeletal motor activity\n",
      "is_a: GO:0140657 ! ATP-dependent activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000147\n",
      "name: actin cortical patch assembly\n",
      "namespace: biological_process\n",
      "def: \"Assembly of an actin cortical patch, a discrete actin-containing structure found at the plasma membrane of fungal cells.\" [GOC:mah]\n",
      "is_a: GO:0022607 ! cellular component assembly\n",
      "is_a: GO:0030866 ! cortical actin cytoskeleton organization\n",
      "is_a: GO:0044396 ! actin cortical patch organization\n",
      "\n",
      "[Term]\n",
      "id: GO:0000148\n",
      "name: 1,3-beta-D-glucan synthase complex\n",
      "namespace: cellular_component\n",
      "def: \"A protein complex that catalyzes the transfer of a glucose group from UDP-glucose to a (1->3)-beta-D-glucan chain.\" [PMID:7983071]\n",
      "synonym: \"(1->3)-beta-glucan synthase complex\" EXACT [GOC:tb]\n",
      "synonym: \"1,3-beta-glucan synthase complex\" EXACT [GOC:tb]\n",
      "is_a: GO:0098797 ! plasma membrane protein complex\n",
      "is_a: GO:1990234 ! transferase complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000149\n",
      "name: SNARE binding\n",
      "namespace: molecular_function\n",
      "def: \"Binding to a SNARE (soluble N-ethylmaleimide-sensitive factor attached protein receptor) protein.\" [PMID:12642621]\n",
      "subset: goslim_chembl\n",
      "synonym: \"SNAP receptor binding\" EXACT []\n",
      "xref: Reactome:R-HSA-210426 \"Glutamate synaptic vesicle docking and priming\"\n",
      "xref: Reactome:R-HSA-210430 \"Release of L-Glutamate at the synapse\"\n",
      "xref: Reactome:R-HSA-265166 \"Exocytosis of Insulin\"\n",
      "xref: Reactome:R-HSA-372505 \"Acetylcholine synaptic vesicle docking and priming\"\n",
      "xref: Reactome:R-HSA-372529 \"Release of acetylcholine at the synapse\"\n",
      "xref: Reactome:R-HSA-374899 \"Release of noradrenaline  at the synapse\"\n",
      "xref: Reactome:R-HSA-374922 \"Noradrenalin synaptic vesicle docking and priming\"\n",
      "xref: Reactome:R-HSA-376357 \"Vamp7 associated Lysosome to Plasma membrane transport\"\n",
      "xref: Reactome:R-HSA-376364 \"Vamp8 associated secretory vesicle to plasma membrane transport\"\n",
      "xref: Reactome:R-HSA-376369 \"Vamp2 associated secretory vesicle to plasma membrane transport\"\n",
      "xref: Reactome:R-HSA-380574 \"Dopamine synaptic vesicle docking and priming\"\n",
      "xref: Reactome:R-HSA-380869 \"Release of docked dopamine loaded synaptic vesicle\"\n",
      "xref: Reactome:R-HSA-380901 \"Release of docked serotonin loaded synaptic vesicle\"\n",
      "xref: Reactome:R-HSA-380905 \"Serotonin loaded synaptic vesicle docking and priming\"\n",
      "xref: Reactome:R-HSA-888589 \"Release of GABA at the synapse\"\n",
      "xref: Reactome:R-HSA-9023173 \"Insulin secretory granule docks at the plasma membrane\"\n",
      "xref: Reactome:R-HSA-917744 \"GABA loaded synaptic vesicle Docking and Priming\"\n",
      "is_a: GO:0005515 ! protein binding\n",
      "\n",
      "[Term]\n",
      "id: GO:0000150\n",
      "name: DNA strand exchange activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the identification and base-pairing of homologous sequences between single-stranded DNA and double-stranded DNA.\" [GOC:elh]\n",
      "comment: Note that this term represents activities that do not break or form phosphodiester bonds, and is therefore not a parent of 'site-specific recombinase activity ; GO:0009009'.\n",
      "subset: goslim_metagenomics\n",
      "synonym: \"RecA-family recombinase activity\" RELATED []\n",
      "synonym: \"recombinase activity\" EXACT []\n",
      "synonym: \"strand exchange activity\" RELATED []\n",
      "synonym: \"strand transferase\" NARROW []\n",
      "xref: Reactome:R-HSA-912458 \"Formation of meiotic heteroduplex\"\n",
      "is_a: GO:0140097 ! catalytic activity, acting on DNA\n",
      "\n",
      "[Term]\n",
      "id: GO:0000151\n",
      "name: ubiquitin ligase complex\n",
      "namespace: cellular_component\n",
      "def: \"A protein complex that includes a ubiquitin-protein ligase and enables ubiquitin protein ligase activity. The complex also contains other proteins that may confer substrate specificity on the complex.\" [GOC:jh2, PMID:9529603]\n",
      "subset: goslim_pir\n",
      "is_a: GO:0140535 ! intracellular protein-containing complex\n",
      "is_a: GO:1990234 ! transferase complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000152\n",
      "name: nuclear ubiquitin ligase complex\n",
      "namespace: cellular_component\n",
      "def: \"A ubiquitin ligase complex found in the nucleus.\" [GOC:mah]\n",
      "is_a: GO:0000151 ! ubiquitin ligase complex\n",
      "is_a: GO:0140513 ! nuclear protein-containing complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000153\n",
      "name: cytoplasmic ubiquitin ligase complex\n",
      "namespace: cellular_component\n",
      "def: \"A ubiquitin ligase complex found in the cytoplasm.\" [GOC:mah]\n",
      "is_a: GO:0000151 ! ubiquitin ligase complex\n",
      "relationship: part_of GO:0005737 ! cytoplasm\n",
      "\n",
      "[Term]\n",
      "id: GO:0000154\n",
      "name: rRNA modification\n",
      "namespace: biological_process\n",
      "alt_id: GO:0016548\n",
      "def: \"The covalent alteration of one or more nucleotides within an rRNA molecule to produce an rRNA molecule with a sequence that differs from that coded genetically.\" [GOC:curators]\n",
      "comment: The term 'RNA editing' (GO:0016547) was merged into 'RNA modification' (GO:0009451) on the basis of statements in the preface of Modification and Editing of RNA (ISBN:1555811337) that there is no clear distinction between modification and editing. Parallel changes were made for substrate (e.g. tRNA, rRNA, etc.) specific child terms of 'RNA editing'.\n",
      "synonym: \"rRNA editing\" NARROW [GOC:hjd]\n",
      "is_a: GO:0006364 ! rRNA processing\n",
      "is_a: GO:0009451 ! RNA modification\n",
      "\n",
      "[Term]\n",
      "id: GO:0000155\n",
      "name: phosphorelay sensor kinase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the phosphorylation of a histidine residue in response to detection of an extracellular signal such as a chemical ligand or change in environment, to initiate a change in cell state or activity. The two-component sensor is a histidine kinase that autophosphorylates a histidine residue in its active site. The phosphate is then transferred to an aspartate residue in a downstream response regulator, to trigger a response.\" [GOC:bf, GOC:mcc, PMID:10966457, PMID:20223701, PMID:9191038]\n",
      "synonym: \"two-component sensor activity\" NARROW []\n",
      "synonym: \"two-component sensor molecule\" NARROW []\n",
      "synonym: \"two-component system sensor activity\" NARROW []\n",
      "xref: EC:2.7.3.-\n",
      "is_a: GO:0004673 ! protein histidine kinase activity\n",
      "is_a: GO:0140299 ! small molecule sensor activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000156\n",
      "name: phosphorelay response regulator activity\n",
      "namespace: molecular_function\n",
      "def: \"Responds to a phosphorelay sensor to initiate a change in cell state or activity. The activity of the response regulator is regulated by transfer of a phosphate from a histidine residue in the sensor, to an aspartate residue in the response regulator. Many but not all response regulators act as transcriptional regulators to elicit a response.\" [GOC:bf, PMID:10966457, PMID:11842140]\n",
      "synonym: \"two-component response regulator activity\" NARROW []\n",
      "is_a: GO:0060089 ! molecular transducer activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000159\n",
      "name: protein phosphatase type 2A complex\n",
      "namespace: cellular_component\n",
      "def: \"A protein complex that has protein serine/threonine phosphatase activity that is polycation-stimulated (PCS), being directly stimulated by protamine, polylysine, or histone H1; it constitutes a subclass of several enzymes activated by different histones and polylysine, and consists of catalytic, scaffolding, and regulatory subunits. The catalytic and scaffolding subunits form the core enzyme, and the holoenzyme also includes the regulatory subunit.\" [GOC:mah, ISBN:0198547684, PMID:17245430]\n",
      "synonym: \"PP2A complex\" EXACT []\n",
      "synonym: \"PP2A-pi\" NARROW []\n",
      "synonym: \"PP2a-protector\" NARROW []\n",
      "synonym: \"protein phosphatase 2 complex\" RELATED [GOC:dph, GOC:rl]\n",
      "is_a: GO:0008287 ! protein serine/threonine phosphatase complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000160\n",
      "name: phosphorelay signal transduction system\n",
      "namespace: biological_process\n",
      "def: \"A conserved series of molecular signals found in prokaryotes and eukaryotes; involves autophosphorylation of a histidine kinase and the transfer of the phosphate group to an aspartate that then acts as a phospho-donor to response regulator proteins.\" [PMID:9191038]\n",
      "subset: goslim_metagenomics\n",
      "synonym: \"histidyl-aspartyl phosphorelay\" EXACT []\n",
      "is_a: GO:0035556 ! intracellular signal transduction\n",
      "\n",
      "[Term]\n",
      "id: GO:0000161\n",
      "name: osmosensory signaling MAPK cascade\n",
      "namespace: biological_process\n",
      "alt_id: GO:0000167\n",
      "alt_id: GO:0000168\n",
      "alt_id: GO:0000169\n",
      "alt_id: GO:0000173\n",
      "alt_id: GO:0007233\n",
      "def: \"The series of molecular signals in which a stress-activated protein kinase (SAPK) cascade relays a signal, containing at least a Hog1/Sty1 family MAPK, a Pbs2/Wis1 family MAPKK and a Ssk2/Win1 family MAP3K.\" [PMID:17604854, PMID:9561267]\n",
      "synonym: \"activation of MAPK activity involved in osmosensory signaling pathway\" NARROW []\n",
      "synonym: \"activation of MAPKK activity during osmolarity sensing\" NARROW []\n",
      "synonym: \"activation of MAPKK activity involved in osmosensory signaling pathway\" NARROW []\n",
      "synonym: \"activation of MAPKKK activity during osmolarity sensing\" NARROW []\n",
      "synonym: \"activation of MAPKKK activity involved in osmosensory signaling pathway\" NARROW []\n",
      "synonym: \"activation of Pbs2 kinase\" NARROW []\n",
      "synonym: \"High Osmolarity Glycerol (HOG) MAPK pathway\" NARROW [PMID:20880736]\n",
      "synonym: \"Hog1 MAPK pathway\" NARROW [PMID:20880736]\n",
      "synonym: \"Hog1/Sty1 stress-activated MAPK cascade\" NARROW []\n",
      "synonym: \"inactivation of MAPK activity involved in osmosensory signaling pathway\" NARROW []\n",
      "synonym: \"MAPK cascade involved in osmosensory signaling pathway\" EXACT []\n",
      "synonym: \"MAPKKK cascade during osmolarity sensing\" EXACT []\n",
      "synonym: \"MAPKKK cascade involved in osmosensory signaling pathway\" EXACT [GOC:signaling]\n",
      "synonym: \"MAPKKK cascade involved in osmosensory signalling pathway\" EXACT [GOC:mah]\n",
      "synonym: \"osmolarity sensing, activation of MAP kinase kinase activity\" NARROW []\n",
      "synonym: \"osmolarity sensing, activation of MAP kinase kinase kinase activity\" NARROW []\n",
      "synonym: \"osmolarity sensing, activation of MAPK activity\" NARROW []\n",
      "synonym: \"osmolarity sensing, activation of MAPKK activity\" NARROW []\n",
      "synonym: \"osmolarity sensing, activation of MAPKKK activity\" NARROW []\n",
      "synonym: \"osmolarity sensing, MAPKKK cascade\" EXACT []\n",
      "synonym: \"termination of MAPK activity during osmolarity sensing\" NARROW []\n",
      "is_a: GO:0007231 ! osmosensory signaling pathway\n",
      "is_a: GO:0051403 ! stress-activated MAPK cascade\n",
      "\n",
      "[Term]\n",
      "id: GO:0000162\n",
      "name: tryptophan biosynthetic process\n",
      "namespace: biological_process\n",
      "alt_id: GO:0009096\n",
      "def: \"The chemical reactions and pathways resulting in the formation of tryptophan, the chiral amino acid 2-amino-3-(1H-indol-3-yl)propanoic acid; tryptophan is synthesized from chorismate via anthranilate.\" [GOC:mah, ISBN:0471331309, MetaCyc:TRPSYN-PWY]\n",
      "synonym: \"aromatic amino acid family biosynthetic process, anthranilate pathway\" EXACT []\n",
      "synonym: \"tryptophan anabolism\" EXACT []\n",
      "synonym: \"tryptophan biosynthesis\" EXACT []\n",
      "synonym: \"tryptophan formation\" EXACT []\n",
      "synonym: \"tryptophan synthesis\" EXACT []\n",
      "xref: MetaCyc:TRPSYN-PWY\n",
      "is_a: GO:0006568 ! tryptophan metabolic process\n",
      "is_a: GO:0009073 ! aromatic amino acid family biosynthetic process\n",
      "is_a: GO:0046219 ! indolalkylamine biosynthetic process\n",
      "is_a: GO:1901607 ! alpha-amino acid biosynthetic process\n",
      "\n",
      "[Term]\n",
      "id: GO:0000164\n",
      "name: protein phosphatase type 1 complex\n",
      "namespace: cellular_component\n",
      "def: \"A protein complex that possesses magnesium-dependent protein serine/threonine phosphatase (AMD phosphatase) activity, and consists of a catalytic subunit and one or more regulatory subunits that dictates the phosphatase's substrate specificity, function, and activity.\" [GOC:mah, GOC:ssd]\n",
      "is_a: GO:0008287 ! protein serine/threonine phosphatase complex\n",
      "relationship: part_of GO:0005737 ! cytoplasm\n",
      "\n",
      "[Term]\n",
      "id: GO:0000165\n",
      "name: MAPK cascade\n",
      "namespace: biological_process\n",
      "alt_id: GO:0007255\n",
      "def: \"An intracellular protein kinase cascade containing at least a MAPK, a MAPKK and a MAP3K. The cascade can also contain an additional tiers: the upstream MAP4K. The kinases in each tier phosphorylate and activate the kinase in the downstream tier to transmit a signal within a cell.\" [GOC:bf, GOC:mtg_signaling_feb11, PMID:20811974, PMID:9561267]\n",
      "comment: MAPK cascades lie downstream of many cell surface receptors and cooperate in transmitting various extracellular signals to the nucleus. One way by which the specificity of each cascade is regulated is through the existence of several distinct components in each tier of the different cascades. The cascades are typically named according to the component in the MAPK tier.\n",
      "synonym: \"ERK/MAPK cascade\" NARROW []\n",
      "synonym: \"MAP kinase cascade\" EXACT []\n",
      "synonym: \"MAP kinase kinase kinase cascade\" EXACT []\n",
      "synonym: \"MAPK signal transduction\" EXACT [GOC:signaling]\n",
      "synonym: \"MAPK signaling\" RELATED []\n",
      "synonym: \"MAPK signalling\" RELATED []\n",
      "synonym: \"MAPKKK cascade\" EXACT []\n",
      "synonym: \"MAPKKK cascade during sporulation\" NARROW []\n",
      "synonym: \"mitogen-activated protein kinase cascade\" EXACT [GOC:bf]\n",
      "xref: Wikipedia:MAPK_cascade\n",
      "xref: Wikipedia:Mitogen-activated_protein_kinase\n",
      "is_a: GO:0035556 ! intracellular signal transduction\n",
      "\n",
      "[Term]\n",
      "id: GO:0000166\n",
      "name: nucleotide binding\n",
      "namespace: molecular_function\n",
      "def: \"Binding to a nucleotide, any compound consisting of a nucleoside that is esterified with (ortho)phosphate or an oligophosphate at any hydroxyl group on the ribose or deoxyribose.\" [GOC:mah, ISBN:0198547684]\n",
      "subset: goslim_chembl\n",
      "subset: goslim_metagenomics\n",
      "subset: goslim_pir\n",
      "subset: goslim_plant\n",
      "is_a: GO:0036094 ! small molecule binding\n",
      "is_a: GO:1901265 ! nucleoside phosphate binding\n",
      "\n",
      "[Term]\n",
      "id: GO:0000170\n",
      "name: sphingosine hydroxylase activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the hydroxylation of sphingolipid long chain bases.\" [PMID:9556590]\n",
      "xref: Reactome:R-HSA-428260 \"dihydroceramide + NADPH + H+ + O2 => phytoceramide + NADP+ + H2O\"\n",
      "is_a: GO:0016491 ! oxidoreductase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000171\n",
      "name: ribonuclease MRP activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the site-specific cleavage of RNA by a catalytic RNA-mediated mechanism; substrates include the A3 site in the ITS1 of pre-rRNA.\" [PMID:17881380]\n",
      "synonym: \"RNase MRP\" EXACT []\n",
      "xref: Wikipedia:RNase_MRP\n",
      "is_a: GO:0004521 ! RNA endonuclease activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000172\n",
      "name: ribonuclease MRP complex\n",
      "namespace: cellular_component\n",
      "def: \"A ribonucleoprotein complex that contains an RNA molecule of the snoRNA family, and cleaves the rRNA precursor as part of rRNA transcript processing. It also has other roles: In S. cerevisiae it is involved in cell cycle-regulated degradation of daughter cell-specific mRNAs, while in mammalian cells it also enters the mitochondria and processes RNAs to create RNA primers for DNA replication.\" [GOC:sgd_curators, PMID:10690410, PMID:14729943, PMID:7510714]\n",
      "synonym: \"ribonuclease mitochondrial RNA processing complex\" EXACT []\n",
      "synonym: \"RNase MRP complex\" EXACT []\n",
      "is_a: GO:0005732 ! sno(s)RNA-containing ribonucleoprotein complex\n",
      "is_a: GO:1902555 ! endoribonuclease complex\n",
      "\n",
      "[Term]\n",
      "id: GO:0000174\n",
      "name: obsolete inactivation of MAPK (mating sensu Saccharomyces)\n",
      "namespace: biological_process\n",
      "def: \"OBSOLETE. Downregulation of MAP kinase activity in the context of transduction of mating pheromone signal, as described for Saccharomyces.\" [PMID:9561267]\n",
      "comment: This term was made obsolete because it is a gene product specific term.\n",
      "synonym: \"inactivation of MAPK (mating sensu Saccharomyces)\" EXACT []\n",
      "is_obsolete: true\n",
      "replaced_by: GO:0000750\n",
      "\n",
      "[Term]\n",
      "id: GO:0000175\n",
      "name: 3'-5'-RNA exonuclease activity\n",
      "namespace: molecular_function\n",
      "def: \"Catalysis of the sequential cleavage of mononucleotides from a free 3' terminus of an RNA molecule.\" [GOC:mah, ISBN:0198547684]\n",
      "synonym: \"3'-5' exoribonuclease activity\" EXACT []\n",
      "synonym: \"3'-5'-exoribonuclease activity\" EXACT []\n",
      "xref: EC:3.1.13.-\n",
      "xref: Reactome:R-HSA-430028 \"Exosome Complex hydrolyzes mRNA by 3' to 5' exoribonuclease digestion\"\n",
      "xref: Reactome:R-HSA-6791222 \"21S pre-rRNA is nucleolytically processed at site E (site2a) to yield 18SE pre-rRNA\"\n",
      "xref: Reactome:R-HSA-6791227 \"47S pre-rRNA is nucleolytically processed at A' (01,A1), site A0, and site 02 (site 6)  to yield 45S pre-rRNA\"\n",
      "xref: Reactome:R-HSA-9682603 \"nsp14 acts as a 3'-to-5' exonuclease to remove misincorporated nucleotides from nascent RNA\"\n",
      "xref: Reactome:R-HSA-9694632 \"nsp14 acts as a 3'-to-5' exonuclease to remove misincorporated nucleotides from nascent RNA\"\n",
      "is_a: GO:0008408 ! 3'-5' exonuclease activity\n",
      "is_a: GO:0016896 ! RNA exonuclease activity, producing 5'-phosphomonoesters\n",
      "\n",
      "[Term]\n",
      "id: GO:0000176\n",
      "name: nuclear exosome (RNase complex)\n",
      "namespace: cellular_component\n",
      "def: \"A ribonuclease complex that has 3-prime to 5-prime processive and distributive hydrolytic exoribonuclease activity and endoribonuclease activity, producing 5-prime-phosphomonoesters. Participates in a multitude of cellular RNA processing and degradation events preventing nuclear export and/or translation of aberrant RNAs. Restricted to processing linear and circular single-stranded RNAs (ssRNA) only. RNAs with complex secondary structures may have to be unwound or pre-processed by co-factors prior to entering the complex, esp if the 3-prime end is structured.\" [PMID:17174896, PMID:20531386, PMID:26726035]\n",
      "synonym: \"eukaryotic exosome multienzyme ribonuclease complex\" EXACT []\n",
      "synonym: \"nuclear exosome (ribonuclease complex)\" EXACT []\n",
      "synonym: \"nuclear exosome multienzyme ribonuclease complex\" EXACT []\n",
      "is_a: GO:0000178 ! exosome (RNase complex)\n",
      "is_a: GO:0140513 ! nuclear protein-containing complex\n",
      "relationship: part_of GO:0031981 ! nuclear lumen\n",
      "\n",
      "[Term]\n",
      "id: GO:0000177\n",
      "name: cytoplasmic exosome (RNase complex)\n",
      "namespace: cellular_component\n",
      "def: \"A ribonuclease complex that has 3-prime to 5-prime processive hydrolytic exoribonuclease activity producing 5-prime-phosphomonoesters. Participates in a multitude of cellular RNA processing and degradation events preventing nuclear export and/or translation of aberrant RNAs. Restricted to processing linear and circular single-stranded RNAs (ssRNA) only. RNAs with complex secondary structures may have to be unwound or pre-processed by co-factors prior to entering the complex, esp if the 3-prime end is structured.\" [PMID:17174896, PMID:20531386, PMID:26726035]\n",
      "synonym: \"cytoplasmic exosome (ribonuclease complex)\" EXACT []\n",
      "synonym: \"cytoplasmic exosome multienzyme ribonuclease complex\" EXACT []\n",
      "synonym: \"prokaryotic exosome multienzyme ribonuclease complex\" EXACT []\n",
      "is_a: GO:0000178 ! exosome (RNase complex)\n",
      "relationship: part_of GO:0005737 ! cytoplasm\n",
      "\n",
      "[Term]\n",
      "id: GO:0000178\n",
      "name: exosome (RNase complex)\n",
      "namespace: cellular_component\n",
      "def: \"A ribonuclease complex that has 3-prime to 5-prime exoribonuclease activity and possibly endoribonuclease activity, producing 5-prime-phosphomonoesters. Participates in a multitude of cellular RNA processing and degradation events preventing nuclear export and/or translation of aberrant RNAs. Restricted to processing linear and circular single-stranded RNAs (ssRNA) only. RNAs with complex secondary structures may have to be unwound or pre-processed by co-factors prior to entering the complex, esp if the 3-prime end is structured.\" [PMID:17174896, PMID:20531386, PMID:26726035]\n",
      "comment: Note that this term should not be confused with 'exosome' used in the context of vesicles released from multivesicular bodies.\n",
      "subset: goslim_pir\n",
      "synonym: \"exosome (ribonucleasease complex)\" EXACT []\n",
      "synonym: \"exosome multienzyme ribonuclease complex\" EXACT []\n",
      "is_a: GO:1905354 ! exoribonuclease complex\n",
      "relationship: part_of GO:0005622 ! intracellular anatomical structure\n",
      "\n",
      "[Term]\n",
      "id: GO:0000179\n",
      "name: rRNA (adenine-N6,N6-)-dimethyltransferase activity\n",
      "namespace: molecular_function\n",
      "alt_id: GO:0043790\n",
      "def: \"Catalysis of the dimethylation of two adjacent adenine residues in a rRNA, using S-adenosyl-L-methionine as a methyl donor.\" [ISBN:1555811337, PMID:10690410]\n",
      "synonym: \"18S rRNA dimethylase activity\" EXACT []\n",
      "synonym: \"dimethyladenosine transferase activity\" EXACT []\n",
      "synonym: \"S-adenosylmethionine-6-N', N'-adenosyl(rRNA) dimethyltransferase activity\" EXACT []\n",
      "xref: Reactome:R-HSA-6790994 \"DIMT1 dimethylates adenosine-1850,1851 of 18S rRNA yielding 6-dimethyladenosine-1850,1851\"\n",
      "xref: Reactome:R-HSA-6793066 \"TFB1M dimethylates adenosine-936 and adenosine-937 of 12S rRNA yielding 6-dimethyladenosine-936 and 6-dimethyladenosine-937\"\n",
      "is_a: GO:0008170 ! N-methyltransferase activity\n",
      "is_a: GO:0016433 ! rRNA (adenine) methyltransferase activity\n",
      "\n",
      "[Term]\n",
      "id: GO:0000180\n",
      "name: obsolete cytosolic large ribosomal subunit\n",
      "namespace: cellular_component\n",
      "def: \"OBSOLETE. (Was not defined before being made obsolete).\" [GOC:ai]\n",
      "comment: This term was made obsolete because more specific children exist.\n",
      "synonym: \"cytosolic large ribosomal subunit\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0022625\n",
      "\n",
      "[Term]\n",
      "id: GO:0000181\n",
      "name: obsolete cytosolic small ribosomal subunit\n",
      "namespace: cellular_component\n",
      "def: \"OBSOLETE. (Was not defined before being made obsolete).\" [GOC:ai]\n",
      "comment: This term was made obsolete because more specific children exist.\n",
      "synonym: \"cytosolic small ribosomal subunit\" EXACT []\n",
      "is_obsolete: true\n",
      "consider: GO:0022627\n",
      "\n",
      "[Term]\n",
      "id: GO:0000182\n",
      "name: rDNA binding\n",
      "namespace: molecular_function\n",
      "def: \"Binding to a DNA sequence encoding a ribosomal RNA\n"
     ]
    }
   ],
   "source": [
    "print(obo_contents[:100000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'format-version': '1.2', 'data-version': 'releases/2023-01-01', 'subsetdef': 'prokaryote_subset \"GO subset for prokaryotes\"', 'synonymtypedef': 'systematic_synonym \"Systematic synonym\" EXACT', 'default-namespace': 'gene_ontology', 'ontology': 'go'}\n",
      "Term 47417\n",
      "Typedef 5\n"
     ]
    }
   ],
   "source": [
    "from typing import Iterable\n",
    "\n",
    "GO_BASIC_MULTIVALUED_KEYS = ['is_a', 'relationship', 'synonym', 'xref', 'alt_id', 'subset', 'consider']\n",
    "\n",
    "\n",
    "def parse_obo_content(obo_lines: Iterable[str], multivalued_keys: Container[str]=GO_BASIC_MULTIVALUED_KEYS) -> dict:\n",
    "    \"\"\"\n",
    "    Parse an OBO file and return a dictionary of header keys and values, and a dictionary of atom types and their values.\n",
    "    Currently defined \"atoms\" in the go-basic.obo file are: \"Term\" and \"Typedef\". The Typedef entries define relationship types between Terms\n",
    "    \"\"\"\n",
    "    def parse_kv(line: str) -> tuple[str, str]:\n",
    "        key, value = line.split(':', 1)\n",
    "        key = key.strip()\n",
    "        value = value.strip()\n",
    "        return key, value\n",
    "    \n",
    "    multivalued_keys = set(multivalued_keys)\n",
    "\n",
    "    header = {}\n",
    "    content = {}\n",
    "    current_atom = {}\n",
    "    current_atom_type = None\n",
    "    for line_num, line in enumerate(obo_lines):\n",
    "        line = line.strip()\n",
    "        # print(line_num, line)\n",
    "        if not line:\n",
    "            continue\n",
    "        if line[0] == '[' and line[-1] == ']':\n",
    "            if current_atom:\n",
    "                # print(f\"Adding new atom of type: {current_atom_type}, {current_atom}\")\n",
    "                content[current_atom_type].append(current_atom)\n",
    "                current_atom = {}\n",
    "            current_atom_type = line[1:-1]\n",
    "            if current_atom_type not in content:\n",
    "                # print(f\"Adding new atom type: {current_atom_type}\")\n",
    "                content[current_atom_type] = []\n",
    "\n",
    "            continue\n",
    "        try:\n",
    "            key, value = parse_kv(line)\n",
    "        except ValueError:\n",
    "            # print(f\"Error parsing line {line_num}: {line}\")\n",
    "            continue\n",
    "        if current_atom_type is None:\n",
    "            # print(f\"Adding header: {key}, {value}\")\n",
    "            header[key] = value\n",
    "        else:\n",
    "            if key in multivalued_keys:\n",
    "                if key not in current_atom:\n",
    "                    current_atom[key] = []\n",
    "                current_atom[key].append(value)\n",
    "            else:\n",
    "                current_atom[key] = value\n",
    "            \n",
    "    if current_atom:\n",
    "        # print(f\"Adding new atom of type: {current_atom_type}, {current_atom}\")\n",
    "        content[current_atom_type].append(current_atom)\n",
    "\n",
    "    return header, content\n",
    "\n",
    "\n",
    "with open('../cafa-5-protein-function-prediction/Train/go-basic.obo', 'r') as f:\n",
    "    header, content = parse_obo_content(f)\n",
    "\n",
    "print(header)\n",
    "for key in content:\n",
    "    print(key, len(content[key]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': 1,\n",
       " 'name': 1,\n",
       " 'namespace': 1,\n",
       " 'def': 1,\n",
       " 'synonym': 284,\n",
       " 'is_a': 11,\n",
       " 'alt_id': 16,\n",
       " 'subset': 12,\n",
       " 'xref': 465,\n",
       " 'comment': 1,\n",
       " 'is_obsolete': 1,\n",
       " 'consider': 8,\n",
       " 'relationship': 3,\n",
       " 'replaced_by': 1}"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_values = {}\n",
    "for term in content['Term']:\n",
    "    for key, values in term.items():\n",
    "        if key not in max_values:\n",
    "            max_values[key] = 0\n",
    "        max_values[key] = max(max_values[key], len(values))\n",
    "\n",
    "max_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': 'GO:0000007',\n",
       "  'name': 'low-affinity zinc ion transmembrane transporter activity',\n",
       "  'namespace': 'molecular_function',\n",
       "  'def': '\"Enables the transfer of a solute or solutes from one side of a membrane to the other according to the reaction: Zn2+ = Zn2+, probably powered by proton motive force. In low-affinity transport the transporter is able to bind the solute only if it is present at very high concentrations.\" [GOC:mtg_transport, ISBN:0815340729]',\n",
       "  'is_a': ['GO:0005385 ! zinc ion transmembrane transporter activity']},\n",
       " {'id': 'GO:0000008',\n",
       "  'name': 'obsolete thioredoxin',\n",
       "  'namespace': 'molecular_function',\n",
       "  'alt_id': ['GO:0000013'],\n",
       "  'def': '\"OBSOLETE. A small disulfide-containing redox protein that serves as a general protein disulfide oxidoreductase. Interacts with a broad range of proteins by a redox mechanism, based on the reversible oxidation of 2 cysteine thiol groups to a disulfide, accompanied by the transfer of 2 electrons and 2 protons. The net result is the covalent interconversion of a disulfide and a dithiol.\" [GOC:kd]',\n",
       "  'comment': 'This term was made obsolete because it represents gene products.',\n",
       "  'synonym': ['\"thioredoxin\" EXACT []'],\n",
       "  'is_obsolete': 'true',\n",
       "  'consider': ['GO:0003756', 'GO:0015036']},\n",
       " {'id': 'GO:0000009',\n",
       "  'name': 'alpha-1,6-mannosyltransferase activity',\n",
       "  'namespace': 'molecular_function',\n",
       "  'def': '\"Catalysis of the transfer of a mannose residue to an oligosaccharide, forming an alpha-(1->6) linkage.\" [GOC:mcc, PMID:2644248]',\n",
       "  'synonym': ['\"1,6-alpha-mannosyltransferase activity\" EXACT []'],\n",
       "  'xref': ['Reactome:R-HSA-449718 \"Addition of a third mannose to the N-glycan precursor by ALG2\"'],\n",
       "  'is_a': ['GO:0000030 ! mannosyltransferase activity']},\n",
       " {'id': 'GO:0000010',\n",
       "  'name': 'trans-hexaprenyltranstransferase activity',\n",
       "  'namespace': 'molecular_function',\n",
       "  'alt_id': ['GO:0036422'],\n",
       "  'def': '\"Catalysis of the reaction: (2E,6E)-farnesyl diphosphate + 4 isopentenyl diphosphate = 4 diphosphate + all-trans-heptaprenyl diphosphate.\" [PMID:9708911, RHEA:27794]',\n",
       "  'synonym': ['\"all-trans-heptaprenyl-diphosphate synthase activity\" RELATED [EC:2.5.1.30]',\n",
       "   '\"HepPP synthase activity\" RELATED [EC:2.5.1.30]',\n",
       "   '\"heptaprenyl diphosphate synthase activity\" RELATED []',\n",
       "   '\"heptaprenyl pyrophosphate synthase activity\" RELATED [EC:2.5.1.30]',\n",
       "   '\"heptaprenyl pyrophosphate synthetase activity\" RELATED [EC:2.5.1.30]'],\n",
       "  'xref': ['EC:2.5.1.30',\n",
       "   'MetaCyc:TRANS-HEXAPRENYLTRANSTRANSFERASE-RXN',\n",
       "   'RHEA:27794'],\n",
       "  'is_a': ['GO:0004659 ! prenyltransferase activity']},\n",
       " {'id': 'GO:0000011',\n",
       "  'name': 'vacuole inheritance',\n",
       "  'namespace': 'biological_process',\n",
       "  'def': '\"The distribution of vacuoles into daughter cells after mitosis or meiosis, mediated by interactions between vacuoles and the cytoskeleton.\" [GOC:mcc, PMID:10873824, PMID:14616069]',\n",
       "  'is_a': ['GO:0007033 ! vacuole organization',\n",
       "   'GO:0048308 ! organelle inheritance']}]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "content['Term'][5:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "def construct_nx_graph_from_obo_content(obo_terms: Iterable[dict], relationship_types_to_include: Container[str]) -> nx.DiGraph:\n",
    "    \"\"\"\n",
    "    Construct a networkx graph from the OBO content.\n",
    "    \"\"\"\n",
    "    def parse_is_a(is_a: str) -> tuple[str, str]:\n",
    "        parent_id, parent_name = is_a.split(' ! ')\n",
    "        return parent_id, parent_name\n",
    "    \n",
    "    def parse_relationship(relationship: str) -> tuple[str, str]:\n",
    "        relationship_type, target = relationship.split(' ')\n",
    "        target_id, target_name = target.split(' ! ')\n",
    "        return relationship_type, target_id, target_name\n",
    "    \n",
    "    valid_relationship_types = set(relationship_types_to_include)\n",
    "\n",
    "    graph = nx.DiGraph()\n",
    "    # Add all non-obsolete terms to the graph\n",
    "    for term in obo_terms:\n",
    "        if 'is_obsolete' in term and term['is_obsolete'] == 'true':\n",
    "            continue\n",
    "        graph.add_node(term['id'], **term)\n",
    "\n",
    "        # Add the \"is_a\" relationships in both parent and child directions\n",
    "        if 'is_a' in term:\n",
    "            for is_a in term['is_a']:\n",
    "                parent_id, _ = parse_is_a(is_a)\n",
    "                graph.add_edge(parent_id, term['id'], relationship_type='CHILD')\n",
    "                graph.add_edge(term['id'], parent_id, relationship_type='PARENT')\n",
    "\n",
    "        # Add the other relationships\n",
    "        if 'relationship' in term:\n",
    "            for relationship in term['relationship']:\n",
    "                relationship_type, target_id, _ = parse_relationship(relationship)\n",
    "                if relationship_type in valid_relationship_types:\n",
    "                    graph.add_edge(term['id'], target_id, relationship_type=relationship_type)\n",
    "    \n",
    "        return graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_go_dag = construct_nx_graph_from_obo_content(content['Term'], set().union([type_def['id'][0] for type_def in content['Typedef']]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<networkx.classes.digraph.DiGraph at 0x164342fd0>"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_go_dag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'negatively_regulates',\n",
       " 'part_of',\n",
       " 'positively_regulates',\n",
       " 'regulates',\n",
       " 'term_tracker_item'}"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set().union([type_def['id'][0] for type_def in content['Typedef']])a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'id': 'GO:0000015', 'name': 'phosphopyruvate hydratase complex', 'namespace': 'cellular_component', 'def': '\"A multimeric enzyme complex, usually a dimer or an octamer, that catalyzes the conversion of 2-phospho-D-glycerate to phosphoenolpyruvate and water.\" [GOC:jl, ISBN:0198506732]', 'subset': 'goslim_metagenomics', 'synonym': '\"enolase complex\" EXACT []', 'is_a': 'GO:1902494 ! catalytic complex', 'relationship': 'part_of GO:0005829 ! cytosol'}\n"
     ]
    }
   ],
   "source": [
    "for term in content['Term']:\n",
    "    if 'relationship' in term:\n",
    "        print(term)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': 'GO:0000007',\n",
       "  'name': 'low-affinity zinc ion transmembrane transporter activity',\n",
       "  'namespace': 'molecular_function',\n",
       "  'def': '\"Enables the transfer of a solute or solutes from one side of a membrane to the other according to the reaction: Zn2+ = Zn2+, probably powered by proton motive force. In low-affinity transport the transporter is able to bind the solute only if it is present at very high concentrations.\" [GOC:mtg_transport, ISBN:0815340729]',\n",
       "  'is_a': 'GO:0005385 ! zinc ion transmembrane transporter activity'},\n",
       " {'id': 'GO:0000008',\n",
       "  'name': 'obsolete thioredoxin',\n",
       "  'namespace': 'molecular_function',\n",
       "  'alt_id': 'GO:0000013',\n",
       "  'def': '\"OBSOLETE. A small disulfide-containing redox protein that serves as a general protein disulfide oxidoreductase. Interacts with a broad range of proteins by a redox mechanism, based on the reversible oxidation of 2 cysteine thiol groups to a disulfide, accompanied by the transfer of 2 electrons and 2 protons. The net result is the covalent interconversion of a disulfide and a dithiol.\" [GOC:kd]',\n",
       "  'comment': 'This term was made obsolete because it represents gene products.',\n",
       "  'synonym': '\"thioredoxin\" EXACT []',\n",
       "  'is_obsolete': 'true',\n",
       "  'consider': 'GO:0015036'},\n",
       " {'id': 'GO:0000009',\n",
       "  'name': 'alpha-1,6-mannosyltransferase activity',\n",
       "  'namespace': 'molecular_function',\n",
       "  'def': '\"Catalysis of the transfer of a mannose residue to an oligosaccharide, forming an alpha-(1->6) linkage.\" [GOC:mcc, PMID:2644248]',\n",
       "  'synonym': '\"1,6-alpha-mannosyltransferase activity\" EXACT []',\n",
       "  'xref': 'Reactome:R-HSA-449718 \"Addition of a third mannose to the N-glycan precursor by ALG2\"',\n",
       "  'is_a': 'GO:0000030 ! mannosyltransferase activity'},\n",
       " {'id': 'GO:0000010',\n",
       "  'name': 'trans-hexaprenyltranstransferase activity',\n",
       "  'namespace': 'molecular_function',\n",
       "  'alt_id': 'GO:0036422',\n",
       "  'def': '\"Catalysis of the reaction: (2E,6E)-farnesyl diphosphate + 4 isopentenyl diphosphate = 4 diphosphate + all-trans-heptaprenyl diphosphate.\" [PMID:9708911, RHEA:27794]',\n",
       "  'synonym': '\"heptaprenyl pyrophosphate synthetase activity\" RELATED [EC:2.5.1.30]',\n",
       "  'xref': 'RHEA:27794',\n",
       "  'is_a': 'GO:0004659 ! prenyltransferase activity'},\n",
       " {'id': 'GO:0000011',\n",
       "  'name': 'vacuole inheritance',\n",
       "  'namespace': 'biological_process',\n",
       "  'def': '\"The distribution of vacuoles into daughter cells after mitosis or meiosis, mediated by interactions between vacuoles and the cytoskeleton.\" [GOC:mcc, PMID:10873824, PMID:14616069]',\n",
       "  'is_a': 'GO:0048308 ! organelle inheritance'}]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "content['Term'][5:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(content['Term'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Term', 'Typedef'}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_lines = 1000000000\n",
    "line_num = 0\n",
    "atom_types = set([])\n",
    "for line in obo_contents.split('\\n'):\n",
    "    line = line.strip()\n",
    "    # print(line)\n",
    "    if not line:\n",
    "        continue\n",
    "    if line[0] == '[' and line[-1] == ']':\n",
    "        atom_types.add(line[1:-1])\n",
    "    line_num += 1\n",
    "    if line_num > max_lines:\n",
    "        break\n",
    "\n",
    "atom_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atom_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30992530"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(go_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ynthesis\" EXACT [GOC:obol]\n",
      "is_a: GO:0018130 ! heterocycle biosynthetic process\n",
      "is_a: GO:0034309 ! primary alcohol biosynthetic process\n",
      "is_a: GO:0042181 ! ketone biosynthetic process\n",
      "is_a: GO:0120255 ! olefinic compound biosynthetic process\n",
      "is_a: GO:1901362 ! organic cyclic compound biosynthetic process\n",
      "is_a: GO:2001316 ! kojic acid metabolic process\n",
      "\n",
      "[Typedef]\n",
      "id: negatively_regulates\n",
      "name: negatively regulates\n",
      "namespace: external\n",
      "xref: RO:0002212\n",
      "is_a: regulates ! regulates\n",
      "\n",
      "[Typedef]\n",
      "id: part_of\n",
      "name: part of\n",
      "namespace: external\n",
      "xref: BFO:0000050\n",
      "is_transitive: true\n",
      "\n",
      "[Typedef]\n",
      "id: positively_regulates\n",
      "name: positively regulates\n",
      "namespace: external\n",
      "xref: RO:0002213\n",
      "holds_over_chain: negatively_regulates negatively_regulates\n",
      "is_a: regulates ! regulates\n",
      "\n",
      "[Typedef]\n",
      "id: regulates\n",
      "name: regulates\n",
      "namespace: external\n",
      "xref: RO:0002211\n",
      "is_transitive: true\n",
      "\n",
      "[Typedef]\n",
      "id: term_tracker_item\n",
      "name: term tracker item\n",
      "namespace: external\n",
      "xref: IAO:0000233\n",
      "is_metadata_tag: true\n",
      "is_class_level: true\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(go_file[-1000:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GOTerm('GO:0000178'):\n",
       "  id:GO:0000178\n",
       "  item_id:GO:0000178\n",
       "  name:exosome (RNase complex)\n",
       "  namespace:cellular_component\n",
       "  _parents: 1 items\n",
       "    GO:1905354\n",
       "  parents: 1 items\n",
       "    GO:1905354\tlevel-03\tdepth-03\texoribonuclease complex [cellular_component]\n",
       "  children: 2 items\n",
       "    GO:0000177\tlevel-05\tdepth-05\tcytoplasmic exosome (RNase complex) [cellular_component]\n",
       "    GO:0000176\tlevel-03\tdepth-05\tnuclear exosome (RNase complex) [cellular_component]\n",
       "  level:4\n",
       "  depth:4\n",
       "  is_obsolete:False\n",
       "  alt_ids: 0 items"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "godag['GO:0000178']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class GODag in module goatools.obo_parser:\n",
      "\n",
      "class GODag(builtins.dict)\n",
      " |  GODag(\n",
      " |      obo_file: str = 'go-basic.obo',\n",
      " |      optional_attrs: Optional[set] = None,\n",
      " |      load_obsolete: bool = False,\n",
      " |      prt=<ipykernel.iostream.OutStream object at 0x10635bf40>\n",
      " |  )\n",
      " |\n",
      " |  Holds the GO DAG as a dict.\n",
      " |\n",
      " |  Method resolution order:\n",
      " |      GODag\n",
      " |      builtins.dict\n",
      " |      builtins.object\n",
      " |\n",
      " |  Methods defined here:\n",
      " |\n",
      " |  __init__(\n",
      " |      self,\n",
      " |      obo_file: str = 'go-basic.obo',\n",
      " |      optional_attrs: Optional[set] = None,\n",
      " |      load_obsolete: bool = False,\n",
      " |      prt=<ipykernel.iostream.OutStream object at 0x10635bf40>\n",
      " |  )\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |\n",
      " |  draw_lineage(\n",
      " |      self,\n",
      " |      recs,\n",
      " |      nodecolor='mediumseagreen',\n",
      " |      edgecolor='lightslateblue',\n",
      " |      dpi=96,\n",
      " |      output='GO_lineage.png',\n",
      " |      engine='pygraphviz',\n",
      " |      gml=False,\n",
      " |      draw_parents=True,\n",
      " |      draw_children=True\n",
      " |  )\n",
      " |      Draw GO DAG subplot.\n",
      " |\n",
      " |  label_wrap(self, label)\n",
      " |      Label text for plot.\n",
      " |\n",
      " |  load_obo_file(self, obo_file, optional_attrs, load_obsolete, prt)\n",
      " |      Read obo file. Store results.\n",
      " |\n",
      " |  make_graph_pydot(\n",
      " |      self,\n",
      " |      recs,\n",
      " |      nodecolor,\n",
      " |      edgecolor,\n",
      " |      dpi,\n",
      " |      format='png',\n",
      " |      draw_parents=True,\n",
      " |      draw_children=True\n",
      " |  )\n",
      " |      draw AMIGO style network, lineage containing one query record.\n",
      " |\n",
      " |  make_graph_pygraphviz(\n",
      " |      self,\n",
      " |      recs,\n",
      " |      nodecolor,\n",
      " |      edgecolor,\n",
      " |      dpi,\n",
      " |      format='png',\n",
      " |      draw_parents=True,\n",
      " |      draw_children=True\n",
      " |  )\n",
      " |      Draw AMIGO style network, lineage containing one query record.\n",
      " |\n",
      " |  paths_to_top(self, term)\n",
      " |      Returns all possible paths to the root node\n",
      " |\n",
      " |      Each path includes the term given. The order of the path is\n",
      " |      top -> bottom, i.e. it starts with the root and ends with the\n",
      " |      given term (inclusively).\n",
      " |\n",
      " |      Parameters:\n",
      " |      -----------\n",
      " |      - term:\n",
      " |          the ID of the GO term, where the paths begin (i.e. the\n",
      " |          accession 'GO:0003682')\n",
      " |\n",
      " |      Returns:\n",
      " |      --------\n",
      " |      - a list of lists of GO Terms\n",
      " |\n",
      " |  query_term(self, term, verbose=False)\n",
      " |      Given a GO ID, return GO object.\n",
      " |\n",
      " |  update_association(self, association)\n",
      " |      Add the GO parents of a gene's associated GO IDs to the gene's association.\n",
      " |\n",
      " |  write_dag(self, out=<ipykernel.iostream.OutStream object at 0x10635bf40>)\n",
      " |      Write info for all GO Terms in obo file, sorted numerically.\n",
      " |\n",
      " |  ----------------------------------------------------------------------\n",
      " |  Static methods defined here:\n",
      " |\n",
      " |  id2int(go_id)\n",
      " |      Given a GO ID, return the int value.\n",
      " |\n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |\n",
      " |  __dict__\n",
      " |      dictionary for instance variables\n",
      " |\n",
      " |  __weakref__\n",
      " |      list of weak references to the object\n",
      " |\n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from builtins.dict:\n",
      " |\n",
      " |  __contains__(self, key, /)\n",
      " |      True if the dictionary has the specified key, else False.\n",
      " |\n",
      " |  __delitem__(self, key, /)\n",
      " |      Delete self[key].\n",
      " |\n",
      " |  __eq__(self, value, /)\n",
      " |      Return self==value.\n",
      " |\n",
      " |  __ge__(self, value, /)\n",
      " |      Return self>=value.\n",
      " |\n",
      " |  __getattribute__(self, name, /)\n",
      " |      Return getattr(self, name).\n",
      " |\n",
      " |  __getitem__(self, key, /)\n",
      " |      Return self[key].\n",
      " |\n",
      " |  __gt__(self, value, /)\n",
      " |      Return self>value.\n",
      " |\n",
      " |  __ior__(self, value, /)\n",
      " |      Return self|=value.\n",
      " |\n",
      " |  __iter__(self, /)\n",
      " |      Implement iter(self).\n",
      " |\n",
      " |  __le__(self, value, /)\n",
      " |      Return self<=value.\n",
      " |\n",
      " |  __len__(self, /)\n",
      " |      Return len(self).\n",
      " |\n",
      " |  __lt__(self, value, /)\n",
      " |      Return self<value.\n",
      " |\n",
      " |  __ne__(self, value, /)\n",
      " |      Return self!=value.\n",
      " |\n",
      " |  __or__(self, value, /)\n",
      " |      Return self|value.\n",
      " |\n",
      " |  __repr__(self, /)\n",
      " |      Return repr(self).\n",
      " |\n",
      " |  __reversed__(self, /)\n",
      " |      Return a reverse iterator over the dict keys.\n",
      " |\n",
      " |  __ror__(self, value, /)\n",
      " |      Return value|self.\n",
      " |\n",
      " |  __setitem__(self, key, value, /)\n",
      " |      Set self[key] to value.\n",
      " |\n",
      " |  __sizeof__(self, /)\n",
      " |      Return the size of the dict in memory, in bytes.\n",
      " |\n",
      " |  clear(self, /)\n",
      " |      Remove all items from the dict.\n",
      " |\n",
      " |  copy(self, /)\n",
      " |      Return a shallow copy of the dict.\n",
      " |\n",
      " |  get(self, key, default=None, /)\n",
      " |      Return the value for key if key is in the dictionary, else default.\n",
      " |\n",
      " |  items(self, /)\n",
      " |      Return a set-like object providing a view on the dict's items.\n",
      " |\n",
      " |  keys(self, /)\n",
      " |      Return a set-like object providing a view on the dict's keys.\n",
      " |\n",
      " |  pop(self, key, default=<unrepresentable>, /)\n",
      " |      D.pop(k[,d]) -> v, remove specified key and return the corresponding value.\n",
      " |\n",
      " |      If the key is not found, return the default if given; otherwise,\n",
      " |      raise a KeyError.\n",
      " |\n",
      " |  popitem(self, /)\n",
      " |      Remove and return a (key, value) pair as a 2-tuple.\n",
      " |\n",
      " |      Pairs are returned in LIFO (last-in, first-out) order.\n",
      " |      Raises KeyError if the dict is empty.\n",
      " |\n",
      " |  setdefault(self, key, default=None, /)\n",
      " |      Insert key with a value of default if key is not in the dictionary.\n",
      " |\n",
      " |      Return the value for key if key is in the dictionary, else default.\n",
      " |\n",
      " |  update(...)\n",
      " |      D.update([E, ]**F) -> None.  Update D from mapping/iterable E and F.\n",
      " |      If E is present and has a .keys() method, then does:  for k in E.keys(): D[k] = E[k]\n",
      " |      If E is present and lacks a .keys() method, then does:  for k, v in E: D[k] = v\n",
      " |      In either case, this is followed by: for k in F:  D[k] = F[k]\n",
      " |\n",
      " |  values(self, /)\n",
      " |      Return an object providing a view on the dict's values.\n",
      " |\n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods inherited from builtins.dict:\n",
      " |\n",
      " |  __class_getitem__(object, /)\n",
      " |      See PEP 585\n",
      " |\n",
      " |  fromkeys(iterable, value=None, /)\n",
      " |      Create a new dictionary with keys from iterable and values set to value.\n",
      " |\n",
      " |  ----------------------------------------------------------------------\n",
      " |  Static methods inherited from builtins.dict:\n",
      " |\n",
      " |  __new__(*args, **kwargs) class method of builtins.dict\n",
      " |      Create and return a new object.  See help(type) for accurate signature.\n",
      " |\n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from builtins.dict:\n",
      " |\n",
      " |  __hash__ = None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(GODag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
